
<!DOCTYPE HTML>
<html lang="" >
    <head>
        <meta charset="UTF-8">
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <title>1-2. MOFA模型-Figure3b · GitBook</title>
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta name="description" content="">
        <meta name="generator" content="GitBook 3.2.3">
        
        
        
    
    <link rel="stylesheet" href="../gitbook/style.css">

    
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-expandable-chapters/expandable-chapters.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-highlight/website.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-search/search.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-fontsettings/website.css">
                
            
        

    

    
        
    
        
    
        
    
        
    
        
    
        
    

        
    
    
    <meta name="HandheldFriendly" content="true"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../gitbook/images/apple-touch-icon-precomposed-152.png">
    <link rel="shortcut icon" href="../gitbook/images/favicon.ico" type="image/x-icon">

    
    <link rel="next" href="gastrulation_3.html" />
    
    
    <link rel="prev" href="gastrulation_1.html" />
    

    </head>
    <body>
        
<div class="book">
    <div class="book-summary">
        
            
<div id="book-search-input" role="search">
    <input type="text" placeholder="Type to search" />
</div>

            
                <nav role="navigation">
                


<ul class="summary">
    
    

    

    
        
        
    
        <li class="chapter " data-level="1.1" data-path="../">
            
                <a href="../">
            
                    
                    Bioinformatics_reproducing
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2" data-path="../">
            
                <a href="../">
            
                    
                    目录
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3" data-path="./">
            
                <a href="./">
            
                    
                    论文1: 单细胞分辨率下小鼠原肠胚形成的多组学分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.3.1" data-path="gastrulation_1.html">
            
                <a href="gastrulation_1.html">
            
                    
                    1-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter active" data-level="1.3.2" data-path="gastrulation_2.html">
            
                <a href="gastrulation_2.html">
            
                    
                    1-2. MOFA模型-Figure3b
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.3" data-path="gastrulation_3.html">
            
                <a href="gastrulation_3.html">
            
                    
                    1-3. 下游分析-Figure3b
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.4" data-path="gastrulation_4.html">
            
                <a href="gastrulation_4.html">
            
                    
                    1-4. 增强子分析-Figure3c
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    

    

    <li class="divider"></li>

    <li>
        <a href="https://www.gitbook.com" target="blank" class="gitbook-link">
            Published with GitBook
        </a>
    </li>
</ul>


                </nav>
            
        
    </div>

    <div class="book-body">
        
            <div class="body-inner">
                
                    

<div class="book-header" role="navigation">
    

    <!-- Title -->
    <h1>
        <i class="fa fa-circle-o-notch fa-spin"></i>
        <a href=".." >1-2. MOFA模型-Figure3b</a>
    </h1>
</div>




                    <div class="page-wrapper" tabindex="-1" role="main">
                        <div class="page-inner">
                            
<div id="book-search-results">
    <div class="search-noresults">
    
                                <section class="normal markdown-section">
                                
                                <h1 id="nature&#x590D;&#x73B0;&#xFF1A;mofa&#x6A21;&#x578B;-figure3b">Nature&#x590D;&#x73B0;&#xFF1A;MOFA&#x6A21;&#x578B;-Figure3b</h1>
<p>&#x5443;&#xFF0C;&#x5176;&#x5B9E;&#x6700;&#x5F00;&#x59CB;&#x662F;&#x60F3;&#x590D;&#x73B0;Figure3c&#x7684;&#x6765;&#x7740;&#xFF0C;&#x7B49;&#x505A;&#x5230;&#x6700;&#x540E;&#x4E00;&#x6B65;&#x7684;&#x65F6;&#x5019;&#x624D;&#x53D1;&#x73B0;&#xFF0C;&#x4F5C;&#x8005;&#x5728;github&#x4E0A;&#x7684;&#x6807;&#x6CE8;&#x51FA;&#x9519;&#x4E86;&#xFF0C;&#x6CA1;&#x529E;&#x6CD5;&#xFF0C;&#x53EA;&#x597D;&#x5C06;&#x9519;&#x5C31;&#x9519;&#x4E86;</p>
<h2 id="1&#x3001;&#x6570;&#x636E;&#x4E0B;&#x8F7D;">1&#x3001;&#x6570;&#x636E;&#x4E0B;&#x8F7D;</h2>
<p>&#x9996;&#x5148;&#x6211;&#x4EEC;&#x9700;&#x8981;&#x628A;&#x4F5C;&#x8005;&#x7684;&#x5904;&#x7406;&#x540E;&#x7684;&#x6570;&#x636E;&#x7ED9;&#x4E0B;&#x8F7D;&#x4E0B;&#x6765;&#xFF0C;&#x603B;&#x5171;33g&#xFF0C;&#x5F53;&#x7136;&#x5206;&#x6790;&#x7684;&#x65F6;&#x5019;&#x4E0D;&#x4F1A;&#x4E00;&#x6B21;&#x8F7D;&#x5165;33g&#x7684;&#x6570;&#x636E;&#x91CF;&#xFF0C;&#x53EA;&#x662F;&#x5176;&#x4E2D;&#x4E00;&#x5C0F;&#x90E8;&#x5206;&#xFF0C;&#x6211;&#x5C06;&#x4E0B;&#x8F7D;&#x597D;&#x7684;&#x5185;&#x5BB9;&#x653E;&#x5230;&#x6587;&#x4EF6;&#x5939;<code>scnmt</code>&#x4E0B;</p>
<pre><code>ftp://ftp.ebi.ac.uk/pub/databases/scnmt_gastrulation/scnmt_gastrulation.tar.gz
</code></pre><p>&#x4F5C;&#x8005;&#x5176;&#x5B9E;&#x6BCF;&#x4E00;&#x5E45;&#x56FE;&#x90FD;&#x7ED9;&#x4E86;&#x4EE3;&#x7801;&#xFF0C;&#x4F46;&#x662F;&#x5728;&#x8FD9;&#x4E2A;&#x6559;&#x7A0B;&#x91CC;&#x9762;&#xFF0C;&#x4E0D;&#x4F1A;&#x5B8C;&#x5168;&#x7167;&#x642C;&#x91CC;&#x9762;&#x7684;&#xFF0C;&#x56E0;&#x4E3A;MOFA&#x8FD9;&#x4E2A;&#x5305;&#x7684;&#x7F18;&#x6545;&#xFF0C;&#x4F60;&#x5927;&#x6982;&#x7387;&#xFF0C;&#x662F;&#x8FD0;&#x884C;&#x4E0D;&#x6210;&#x529F;&#x7684;&#xFF0C;&#x4F46;&#x8FD9;&#x91CC;&#x8FD8;&#x662F;&#x653E;&#x51FA;&#x4F5C;&#x8005;&#x4EE3;&#x7801;&#x7684;&#x4ED3;&#x5E93;&#xFF0C;&#x5EFA;&#x8BAE;&#x4E00;&#x5E76;&#x4E0B;&#x8F7D;&#x4E0B;&#x6765;&#xFF0C;&#x53EF;&#x4EE5;&#x4F9B;&#x53C2;&#x8003;</p>
<pre><code>https://github.com/rargelaguet/scnmt_gastrulation
</code></pre><h2 id="2&#x3001;mofa&#x6A21;&#x578B;&#x6570;&#x636E;&#x51C6;&#x5907;">2&#x3001;MOFA&#x6A21;&#x578B;&#x6570;&#x636E;&#x51C6;&#x5907;</h2>
<h3 id="21-&#x73AF;&#x5883;&#x51C6;&#x5907;">2.1 &#x73AF;&#x5883;&#x51C6;&#x5907;</h3>
<p>&#x6211;&#x4EEC;&#x5728;&#x521A;&#x521A;&#x5B89;&#x88C5;&#x597D;&#x7684;wsl&#x7CFB;&#x7EDF;&#x91CC;&#x9762;&#xFF0C;&#x8F93;&#x5165;jupyter lab&#xFF0C;&#x6253;&#x5F00;R&#x73AF;&#x5883;&#xFF0C;&#x53EF;&#x4EE5;&#x770B;&#x5230;&#x53F3;&#x4E0A;&#x89D2;&#x6709;&#x4E2A;R&#x7684;&#x6807;&#x6CE8;</p>
<p><img src="https://raw.githubusercontent.com/Starlitnightly/bioinformatic_galaxy/master/img/image-20211006231647297.png" alt="image-20211006231647297"></p>
<h3 id="22-&#x8F7D;&#x5165;r&#x5305;">2.2 &#x8F7D;&#x5165;R&#x5305;</h3>
<p>&#x63A5;&#x4E0B;&#x6765;&#x5C31;&#x662F;&#x6B63;&#x5F0F;&#x7684;&#x8DD1;&#x4EE3;&#x7801;&#x73AF;&#x8282;&#xFF0C;&#x4E3A;&#x4EC0;&#x4E48;&#x662F;wsl&#x5462;&#xFF0C;&#x56E0;&#x4E3A;&#x6709;&#x4E00;&#x4E9B;&#x4EE3;&#x7801;&#x6211;&#x6682;&#x65F6;&#x8FD8;&#x4E0D;&#x77E5;&#x9053;&#x600E;&#x4E48;&#x4FEE;&#x6539;&#xFF0C;&#x53EA;&#x80FD;&#x4FDD;&#x6301;&#x8DDF;&#x4F5C;&#x8005;&#x4E00;&#x81F4;&#x7684;&#x4EE3;&#x7801;</p>
<p>&#x4EE5;&#x4E0B;#&#x4EE3;&#x8868;&#x6E90;&#x6CE8;&#x91CA;&#xFF0C;##&#x4EE3;&#x8868;&#x6211;&#x7684;&#x6CE8;&#x91CA;</p>
<pre><code class="lang-R">suppressMessages(<span class="hljs-keyword">library</span>(MOFA2))
suppressMessages(<span class="hljs-keyword">library</span>(data.table))
suppressMessages(<span class="hljs-keyword">library</span>(purrr))
suppressMessages(<span class="hljs-keyword">library</span>(ggplot2))
suppressMessages(<span class="hljs-keyword">library</span>(scater))
suppressMessages(<span class="hljs-keyword">library</span>(reticulate))
suppressMessages(<span class="hljs-keyword">library</span>(argparse))
</code></pre>
<h3 id="23-&#x914D;&#x7F6E;io">2.3 &#x914D;&#x7F6E;I/O</h3>
<p>&#x5728;&#x8FD9;&#x4E00;&#x6B65;&#xFF0C;&#x53EA;&#x9700;&#x8981;&#x6CE8;&#x610F;<code>io$basedir &lt;- &quot;/mnt/d/drive/scnmt&quot;</code>&#x7684;&#x5185;&#x5BB9;&#x5C31;&#x597D;&#x4E86;&#xFF0C;&#x628A;&#x8FD9;&#x4E2A;&#x6539;&#x6210;&#x4F60;&#x4E0A;&#x9762;&#x6570;&#x636E;&#x4E0B;&#x8F7D;&#x4E0B;&#x6765;&#x7684;&#x5B89;&#x88C5;&#x8DEF;&#x5F84;</p>
<pre><code class="lang-R"><span class="hljs-comment">## Define I/O ##</span>

io$basedir &lt;- <span class="hljs-string">&quot;/mnt/d/drive/scnmt&quot;</span>
io$gene_metadata &lt;- <span class="hljs-string">&quot;/mnt/d/drive/scnmt/gene/Mmusculus_genes_BioMart.87.txt&quot;</span>
io$outdir &lt;- <span class="hljs-string">&quot;/mnt/d/drive/scnmt/result&quot;</span>

io$sample.metadata &lt;- paste0(io$basedir,<span class="hljs-string">&quot;/sample_metadata.txt&quot;</span>)
io$met.dir &lt;- paste0(io$basedir,<span class="hljs-string">&quot;/met/feature_level&quot;</span>)
io$acc.dir &lt;- paste0(io$basedir,<span class="hljs-string">&quot;/acc/feature_level&quot;</span>)
io$rna.file &lt;- paste0(io$basedir,<span class="hljs-string">&quot;/rna/SingleCellExperiment.rds&quot;</span>)
io$annos_dir  &lt;- paste0(io$basedir, <span class="hljs-string">&quot;/features/genomic_contexts&quot;</span>)

<span class="hljs-comment"># Folders with the global statistics per cell</span>
io$met.stats &lt;- paste0(io$basedir,<span class="hljs-string">&quot;/met/results/stats/sample_stats.txt&quot;</span>)
io$acc.stats &lt;- paste0(io$basedir,<span class="hljs-string">&quot;/acc/results/stats/sample_stats.txt&quot;</span>)
</code></pre>
<h3 id="24-&#x5B9A;&#x4E49;&#x53D8;&#x91CF;">2.4 &#x5B9A;&#x4E49;&#x53D8;&#x91CF;</h3>
<p>&#x5728;&#x8FD9;&#x91CC;&#xFF0C;&#x6211;&#x4EEC;&#x4F1A;&#x628A;&#x672C;&#x6B21;&#x6570;&#x636E;&#x5206;&#x6790;&#x6240;&#x9700;&#x8981;&#x7684;&#x53D8;&#x91CF;&#x4EE5;&#x53CA;&#x73AF;&#x5883;&#x90FD;&#x5B9A;&#x4E49;&#x6E05;&#x695A;</p>
<pre><code class="lang-R"><span class="hljs-comment">## Define options ##</span>
opts &lt;- list()

<span class="hljs-comment"># Define which annotations to look at</span>
opts$met.annos &lt;- c(
  <span class="hljs-comment"># &quot;genebody&quot;,</span>
  <span class="hljs-string">&quot;prom_2000_2000&quot;</span>,
  <span class="hljs-string">&quot;H3K27ac_distal_E7.5_Mes_intersect12&quot;</span>,
  <span class="hljs-string">&quot;H3K27ac_distal_E7.5_Ect_intersect12&quot;</span>,
  <span class="hljs-string">&quot;H3K27ac_distal_E7.5_End_intersect12&quot;</span>
  <span class="hljs-comment"># &quot;H3K4me3_E7.5_Mes&quot;,</span>
  <span class="hljs-comment"># &quot;H3K4me3_E7.5_End&quot;,</span>
  <span class="hljs-comment"># &quot;H3K4me3_E7.5_Ect&quot;,</span>
)

opts$acc.annos &lt;- c(
  <span class="hljs-comment"># &quot;genebody&quot;,</span>
  <span class="hljs-string">&quot;prom_2000_2000&quot;</span>,
  <span class="hljs-string">&quot;H3K27ac_distal_E7.5_Mes_intersect12&quot;</span>,
  <span class="hljs-string">&quot;H3K27ac_distal_E7.5_Ect_intersect12&quot;</span>,
  <span class="hljs-string">&quot;H3K27ac_distal_E7.5_End_intersect12&quot;</span>
  <span class="hljs-comment"># &quot;H3K4me3_E7.5_Mes&quot;,</span>
  <span class="hljs-comment"># &quot;H3K4me3_E7.5_End&quot;,</span>
  <span class="hljs-comment"># &quot;H3K4me3_E7.5_Ect&quot;,</span>
)


<span class="hljs-comment"># Define which stage and lineages to look at </span>
opts$stage_lineage &lt;- c(

  <span class="hljs-comment"># E4.5</span>
  <span class="hljs-string">&quot;E4.5_Epiblast&quot;</span>,

  <span class="hljs-comment"># E5.5</span>
  <span class="hljs-string">&quot;E5.5_Epiblast&quot;</span>,

  <span class="hljs-comment"># E6.5</span>
  <span class="hljs-string">&quot;E6.5_Epiblast&quot;</span>,
  <span class="hljs-string">&quot;E6.5_Primitive_Streak&quot;</span>,
  <span class="hljs-string">&quot;E6.5_Mesoderm&quot;</span>,

  <span class="hljs-comment"># E7.5</span>
  <span class="hljs-string">&quot;E7.5_Epiblast&quot;</span>,
  <span class="hljs-string">&quot;E7.5_Primitive_Streak&quot;</span>,
  <span class="hljs-string">&quot;E7.5_Ectoderm&quot;</span>,
  <span class="hljs-string">&quot;E7.5_Endoderm&quot;</span>,
  <span class="hljs-string">&quot;E7.5_Mesoderm&quot;</span>
)

<span class="hljs-comment"># Filtering options for methylation</span>
opts$met_min.CpGs &lt;- <span class="hljs-number">1</span>        <span class="hljs-comment"># minimum number of CpG sites per feature</span>
opts$met_min.cells &lt;- <span class="hljs-number">25</span>      <span class="hljs-comment"># minimum number of cells per feature</span>
opts$met_nfeatures &lt;- <span class="hljs-number">1000</span>    <span class="hljs-comment"># maximum number of features per view (filter based on variance)</span>

<span class="hljs-comment"># Filtering options for accessibility</span>
opts$acc_min.GpCs &lt;- <span class="hljs-number">5</span>        <span class="hljs-comment"># minimum number of GpC sites per feature</span>
opts$acc_min.cells &lt;- <span class="hljs-number">25</span>      <span class="hljs-comment"># minimum number of cells per feature</span>
opts$acc_nfeatures &lt;- <span class="hljs-number">1000</span>    <span class="hljs-comment"># maximum number of features per view (filter based on variance)</span>

<span class="hljs-comment"># Filtering options for RNA</span>
opts$rna_min.cdr &lt;- <span class="hljs-number">0.25</span>      <span class="hljs-comment"># Remove genes with cellular detection rate smaller than opts$min.cdr</span>
opts$rna_ngenes &lt;- <span class="hljs-number">2500</span>       <span class="hljs-comment"># maximum number of genes (filter based on variance)</span>

<span class="hljs-comment"># Define colors</span>
opts$colors &lt;- c(
  <span class="hljs-string">&quot;Epiblast&quot;</span>=<span class="hljs-string">&quot;grey70&quot;</span>,
  <span class="hljs-string">&quot;Mesoderm&quot;</span>=<span class="hljs-string">&quot;#CD3278&quot;</span>,
  <span class="hljs-string">&quot;Primitive Streak&quot;</span>=<span class="hljs-string">&quot;sandybrown&quot;</span>,
  <span class="hljs-string">&quot;Endoderm&quot;</span>=<span class="hljs-string">&quot;#43CD80&quot;</span>,
  <span class="hljs-string">&quot;Ectoderm&quot;</span>=<span class="hljs-string">&quot;steelblue&quot;</span>
)

<span class="hljs-comment"># window length for the overlap between genes and features</span>
opts$overlapGenes  &lt;- <span class="hljs-literal">FALSE</span>
opts$gene_window  &lt;- <span class="hljs-number">5e4</span>

<span class="hljs-comment"># Define which cells to use</span>
<span class="hljs-comment">##&#x4ECE;sample.metadata&#x91CC;&#x9762;&#x8BFB;&#x53D6;&#x4E86;&#x6837;&#x672C;&#x6570;&#x636E;&#xFF0C;&#x5E76;&#x4E14;&#x505A;&#x4E86;&#x4EE5;&#x4E0B;&#x5904;&#x7406;</span>
<span class="hljs-comment">##1&#x3001;&#x5B9A;&#x4E49;&#x4E86;&#x65B0;&#x5217;stage_lineage&#xFF0C;&#x7C7B;&#x578B;&#x4E3A;factor&#xFF0C;&#x503C;&#x4E3A;stage&#x4E0E;lineage10x_2&#x5217;&#x7684;&#x503C;&#x62FC;&#x63A5;&#x800C;&#x6210;</span>
<span class="hljs-comment">##2&#x3001;&#x79FB;&#x9664;&#x4E86;&#x4E0D;&#x5728;opts$stage_lineage&#x91CC;&#x7684;stage_lineage&#x7684;&#x884C;</span>
tmp &lt;- fread(io$sample.metadata) %&gt;%
  .[,stage_lineage:=as.factor(paste(stage,lineage10x_2,sep=<span class="hljs-string">&quot;_&quot;</span>))] %&gt;%
  .[stage_lineage%<span class="hljs-keyword">in</span>%opts$stage_lineage]
<span class="hljs-comment">##&#x9009;&#x62E9;&#x4E86;pass_metQC&#x7B49;&#x4E8E;True&#x7684;&#x884C;&#xFF0C;&#x5E76;&#x5355;&#x72EC;&#x53D6;&#x51FA;id_met&#x5217;&#x5B58;&#x8FDB;met_cell&#x5217;&#x8868;&#x91CC;&#x9762;</span>
opts$met_cells &lt;- tmp %&gt;% .[pass_metQC==<span class="hljs-literal">T</span>, id_met]
<span class="hljs-comment">##&#x9009;&#x62E9;&#x4E86;pass_rnaQC&#x7B49;&#x4E8E;True&#x7684;&#x884C;&#xFF0C;&#x5E76;&#x5355;&#x72EC;&#x53D6;&#x51FA;id_rna&#x5217;</span>
opts$rna_cells &lt;- tmp %&gt;% .[pass_rnaQC==<span class="hljs-literal">T</span>, id_rna]
opts$acc_cells &lt;- tmp %&gt;% .[pass_accQC==<span class="hljs-literal">T</span>, id_acc]
</code></pre>
<h3 id="25-&#x5B9A;&#x4E49;&#x5355;&#x7EC6;&#x80DE;&#x4FE1;&#x606F;">2.5 &#x5B9A;&#x4E49;&#x5355;&#x7EC6;&#x80DE;&#x4FE1;&#x606F;</h3>
<p>&#x6BCF;&#x4E2A;&#x5355;&#x7EC6;&#x80DE;&#x7684;&#x7279;&#x5F81;&#x4F5C;&#x8005;&#x5DF2;&#x7ECF;&#x5904;&#x7406;&#x597D;&#x4E86;&#xFF0C;&#x6211;&#x4EEC;&#x8FD9;&#x91CC;&#x53EA;&#x9700;&#x8981;&#x76F4;&#x63A5;&#x5BFC;&#x5165;&#x5C31;&#x597D;</p>
<pre><code class="lang-R"><span class="hljs-comment">##########################</span>
<span class="hljs-comment">## Load sample metadata ##</span>
<span class="hljs-comment">##########################</span>

sample_metadata &lt;- fread(io$sample.metadata) %&gt;%
  .[,stage_lineage:=paste(stage,lineage10x_2,sep=<span class="hljs-string">&quot;_&quot;</span>)] %&gt;%
  .[id_met%<span class="hljs-keyword">in</span>%opts$met_cells | id_rna %<span class="hljs-keyword">in</span>% opts$rna_cells | id_acc %<span class="hljs-keyword">in</span>% opts$acc_cells ]
</code></pre>
<h3 id="26-&#x5B9A;&#x4E49;&#x77E9;&#x9635;&#x5904;&#x7406;&#x51FD;&#x6570;">2.6 &#x5B9A;&#x4E49;&#x77E9;&#x9635;&#x5904;&#x7406;&#x51FD;&#x6570;</h3>
<pre><code class="lang-R">matrix.please &lt;- <span class="hljs-keyword">function</span>(x) {
  m&lt;-as.matrix(x[,-<span class="hljs-number">1</span>])
  rownames(m)&lt;-x[[<span class="hljs-number">1</span>]]
  m
}
</code></pre>
<h3 id="27-&#x52A0;&#x8F7D;&#x6570;&#x636E;">2.7 &#x52A0;&#x8F7D;&#x6570;&#x636E;</h3>
<p>&#x8FD9;&#x4E00;&#x6B65;&#x8981;&#x7528;&#x5230;linux&#x547D;&#x4EE4;zcat&#xFF0C;&#x6211;&#x4E0D;&#x662F;&#x5F88;&#x4F1A;&#x6539;&#x6210;windows&#x4E0B;&#x7684;&#x547D;&#x4EE4;&#xFF0C;&#x4E0D;&#x7136;&#x6211;&#x5C31;&#x76F4;&#x63A5;&#x6539;&#x4E86;&#xFF0C;&#x4E5F;&#x5C31;&#x4E0D;&#x7528;&#x5B89;wsl&#x4E86;</p>
<pre><code class="lang-R"><span class="hljs-comment"># Load Methylation data</span>
met_dt &lt;- lapply(opts$met.annos, <span class="hljs-keyword">function</span>(n) {
  data &lt;- fread(cmd=sprintf(<span class="hljs-string">&quot;zcat &lt; %s/%s.tsv.gz&quot;</span>,io$met.dir,n), showProgress=<span class="hljs-literal">F</span>, stringsAsFactors=<span class="hljs-literal">F</span>, quote=<span class="hljs-string">&quot;&quot;</span>) %&gt;% .[V1%<span class="hljs-keyword">in</span>%opts$met_cells]
}) %&gt;% rbindlist %&gt;% setnames(c(<span class="hljs-string">&quot;id_met&quot;</span>,<span class="hljs-string">&quot;id&quot;</span>,<span class="hljs-string">&quot;anno&quot;</span>,<span class="hljs-string">&quot;Nmet&quot;</span>,<span class="hljs-string">&quot;N&quot;</span>,<span class="hljs-string">&quot;rate&quot;</span>))

<span class="hljs-comment"># Load Accessibility data</span>
acc_dt &lt;- lapply(opts$acc.annos, <span class="hljs-keyword">function</span>(n) {
  data &lt;- fread(cmd=sprintf(<span class="hljs-string">&quot;zcat &lt; %s/%s.tsv.gz&quot;</span>,io$acc.dir,n), showProgress=<span class="hljs-literal">F</span>, stringsAsFactors=<span class="hljs-literal">F</span>, quote=<span class="hljs-string">&quot;&quot;</span>) %&gt;%
    .[V1%<span class="hljs-keyword">in</span>%opts$acc_cells]
}) %&gt;% rbindlist %&gt;% setnames(c(<span class="hljs-string">&quot;id_acc&quot;</span>,<span class="hljs-string">&quot;id&quot;</span>,<span class="hljs-string">&quot;anno&quot;</span>,<span class="hljs-string">&quot;Nmet&quot;</span>,<span class="hljs-string">&quot;N&quot;</span>,<span class="hljs-string">&quot;rate&quot;</span>))

<span class="hljs-comment"># Load RNA data</span>
sce &lt;- readRDS(io$rna.file) %&gt;% .[,opts$rna_cells]

<span class="hljs-comment"># Load annotation metadata</span>
feature_metadata &lt;- lapply(unique(c(opts$met.annos,opts$acc.annos)), <span class="hljs-keyword">function</span>(i) 
  fread(sprintf(<span class="hljs-string">&quot;%s/%s.bed&quot;</span>,io$annos_dir,i), stringsAsFactors=<span class="hljs-literal">T</span>)[,c(<span class="hljs-number">1</span>,<span class="hljs-number">2</span>,<span class="hljs-number">3</span>,<span class="hljs-number">4</span>,<span class="hljs-number">5</span>,<span class="hljs-number">6</span>)]) %&gt;%
  rbindlist %&gt;% setnames(c(<span class="hljs-string">&quot;chr&quot;</span>,<span class="hljs-string">&quot;start&quot;</span>,<span class="hljs-string">&quot;end&quot;</span>,<span class="hljs-string">&quot;strand&quot;</span>,<span class="hljs-string">&quot;id&quot;</span>,<span class="hljs-string">&quot;anno&quot;</span>))


<span class="hljs-comment"># Load gene metadata </span>
gene_metadata &lt;- fread(io$gene_metadata,stringsAsFactors=<span class="hljs-literal">T</span>) %&gt;% 
  setnames(c(<span class="hljs-string">&quot;ens_id&quot;</span>,<span class="hljs-string">&quot;symbol&quot;</span>),c(<span class="hljs-string">&quot;id&quot;</span>,<span class="hljs-string">&quot;gene&quot;</span>)) %&gt;% 
  .[,chr:=as.factor(sub(<span class="hljs-string">&quot;chr&quot;</span>,<span class="hljs-string">&quot;&quot;</span>,chr))]
</code></pre>
<h3 id="28-&#x6570;&#x636E;&#x89E3;&#x6790;">2.8 &#x6570;&#x636E;&#x89E3;&#x6790;</h3>
<p>&#x4F5C;&#x8005;&#x5BF9;&#x6570;&#x636E;&#x90FD;&#x8FDB;&#x884C;&#x4E86;&#x7B80;&#x5355;&#x7684;&#x89E3;&#x6790;</p>
<pre><code class="lang-R"><span class="hljs-comment"># Parse gene and feature metadata</span>
feature_metadata_filt.met &lt;- feature_metadata %&gt;% split(.$anno) %&gt;% 
  map2(.,names(.), <span class="hljs-keyword">function</span>(x,y) x[id %<span class="hljs-keyword">in</span>% met_dt[anno==y,id]] ) %&gt;%
  rbindlist
feature_metadata_filt.acc &lt;- feature_metadata %&gt;% split(.$anno) %&gt;% 
  map2(.,names(.), <span class="hljs-keyword">function</span>(x,y) x[id %<span class="hljs-keyword">in</span>% acc_dt[anno==y,id]] ) %&gt;%
  rbindlist
gene_metadata_filt &lt;- gene_metadata %&gt;% .[,c(<span class="hljs-string">&quot;chr&quot;</span>,<span class="hljs-string">&quot;start&quot;</span>,<span class="hljs-string">&quot;end&quot;</span>,<span class="hljs-string">&quot;gene&quot;</span>)] %&gt;% 
  .[,c(<span class="hljs-string">&quot;start&quot;</span>, <span class="hljs-string">&quot;end&quot;</span>) := list(start-opts$gene_window, end+opts$gene_window)] %&gt;% 
  setkey(chr,start,end)

<span class="hljs-comment">## Parse RNA expression data ##</span>

<span class="hljs-comment"># Convert to data.table</span>
rna_dt &lt;- exprs(sce) %&gt;% t %&gt;% as.data.table(keep.rownames = <span class="hljs-string">&quot;id_rna&quot;</span>) %&gt;% 
  melt(id.vars = <span class="hljs-string">&quot;id_rna&quot;</span>, value.name = <span class="hljs-string">&quot;expr&quot;</span>, variable.name = <span class="hljs-string">&quot;ens_id&quot;</span>) %&gt;%
  merge(rowData(sce) %&gt;% as.data.frame(row.names = rownames(sce)) %&gt;% tibble::rownames_to_column(<span class="hljs-string">&quot;ens_id&quot;</span>) %&gt;% .[,c(<span class="hljs-string">&quot;symbol&quot;</span>,<span class="hljs-string">&quot;ens_id&quot;</span>)] %&gt;% setnames(<span class="hljs-string">&quot;symbol&quot;</span>,<span class="hljs-string">&quot;gene&quot;</span>))

<span class="hljs-comment">## Parse accessibility data ##</span>
acc_dt[,m:=log2(((rate/<span class="hljs-number">100</span>)+<span class="hljs-number">0.01</span>)/(<span class="hljs-number">1</span>-(rate/<span class="hljs-number">100</span>)+<span class="hljs-number">0.01</span>))] <span class="hljs-comment"># Calculate M value from Beta value</span>

<span class="hljs-comment">## Parse methylation data ##</span>
met_dt[,m:=log2(((rate/<span class="hljs-number">100</span>)+<span class="hljs-number">0.01</span>)/(<span class="hljs-number">1</span>-(rate/<span class="hljs-number">100</span>)+<span class="hljs-number">0.01</span>))] <span class="hljs-comment"># Calculate M value from Beta value</span>

<span class="hljs-comment">##############################</span>
<span class="hljs-comment">## Merge data with metadata ##</span>
<span class="hljs-comment">##############################</span>

acc_dt &lt;- merge(acc_dt, sample_metadata[,c(<span class="hljs-string">&quot;sample&quot;</span>,<span class="hljs-string">&quot;id_acc&quot;</span>,<span class="hljs-string">&quot;stage&quot;</span>,<span class="hljs-string">&quot;stage_lineage&quot;</span>)], by=<span class="hljs-string">&quot;id_acc&quot;</span>) %&gt;% droplevels()
met_dt &lt;- merge(met_dt, sample_metadata[,c(<span class="hljs-string">&quot;sample&quot;</span>,<span class="hljs-string">&quot;id_met&quot;</span>,<span class="hljs-string">&quot;stage&quot;</span>,<span class="hljs-string">&quot;stage_lineage&quot;</span>)], by=<span class="hljs-string">&quot;id_met&quot;</span>) %&gt;% droplevels()
rna_dt &lt;- merge(rna_dt, sample_metadata[,c(<span class="hljs-string">&quot;sample&quot;</span>,<span class="hljs-string">&quot;id_rna&quot;</span>,<span class="hljs-string">&quot;stage&quot;</span>,<span class="hljs-string">&quot;stage_lineage&quot;</span>)], by=<span class="hljs-string">&quot;id_rna&quot;</span>) %&gt;% droplevels()
</code></pre>
<h3 id="29-&#x5C06;&#x57FA;&#x56E0;&#x7EC4;&#x7279;&#x5F81;&#x4E0E;&#x91CD;&#x53E0;&#x57FA;&#x56E0;&#x8054;&#x7CFB;&#x8D77;&#x6765;">2.9 &#x5C06;&#x57FA;&#x56E0;&#x7EC4;&#x7279;&#x5F81;&#x4E0E;&#x91CD;&#x53E0;&#x57FA;&#x56E0;&#x8054;&#x7CFB;&#x8D77;&#x6765;</h3>
<pre><code class="lang-R"><span class="hljs-comment"># Methylation</span>
<span class="hljs-keyword">if</span> (opts$overlapGenes) {
  met_list &lt;- list()
  <span class="hljs-keyword">for</span> (i <span class="hljs-keyword">in</span> unique(met_dt$anno)){

    <span class="hljs-comment"># Subset corresponding anno</span>
    met_tmp &lt;- met_dt[anno==i, ]

    <span class="hljs-comment"># Non gene-associated feature</span>
    <span class="hljs-keyword">if</span> (all(grepl(<span class="hljs-string">&quot;ENSMUSG&quot;</span>, unique(met_tmp$id)) == <span class="hljs-literal">FALSE</span>)) {

      <span class="hljs-comment"># Extract coordiantes for methylation sites and for genes</span>
      feature_metadata_tmp &lt;- feature_metadata_filt.met[anno==i, c(<span class="hljs-string">&quot;chr&quot;</span>,<span class="hljs-string">&quot;start&quot;</span>,<span class="hljs-string">&quot;end&quot;</span>,<span class="hljs-string">&quot;id&quot;</span>)] %&gt;% 
        .[,c(<span class="hljs-string">&quot;start&quot;</span>,<span class="hljs-string">&quot;end&quot;</span>) := list(start - opts$gene_window, end + opts$gene_window)] %&gt;% setkey(chr,start,end)

      <span class="hljs-comment"># Do the overlap</span>
      ov &lt;- foverlaps(
        gene_metadata_filt, 
        feature_metadata_tmp, 
        nomatch=<span class="hljs-number">0</span>) %&gt;% .[,c(<span class="hljs-string">&quot;gene&quot;</span>, <span class="hljs-string">&quot;id&quot;</span>)]

      <span class="hljs-comment"># If a feature overlaps with multiple genes, collapse them</span>
      ov1 &lt;- ov[is.na(gene)]
      ov2 &lt;- ov[!is.na(gene)] %&gt;% .[,.(gene=paste(gene,collapse=<span class="hljs-string">&quot;_&quot;</span>)), by=<span class="hljs-string">&quot;id&quot;</span>]
      ov &lt;- rbind(ov1,ov2)

      <span class="hljs-comment"># Merge with methylation data</span>
      met_list[[i]] &lt;- merge(met_tmp, ov, by=<span class="hljs-string">&quot;id&quot;</span>, allow.cartesian=<span class="hljs-literal">T</span>) 
    }
    <span class="hljs-comment"># Gene-associated feature</span>
    <span class="hljs-keyword">else</span> <span class="hljs-keyword">if</span> (all(grepl(<span class="hljs-string">&quot;ENSMUSG&quot;</span>, unique(met_tmp$id)) == <span class="hljs-literal">TRUE</span>)) {
      met_list[[i]] &lt;- merge(met_tmp, gene_metadata[,c(<span class="hljs-string">&quot;id&quot;</span>,<span class="hljs-string">&quot;gene&quot;</span>)], by=<span class="hljs-string">&quot;id&quot;</span>)
    }
  }
  met_dt &lt;- rbindlist(met_list)
  rm(met_list, met_tmp,feature_metadata_tmp,ov)
} <span class="hljs-keyword">else</span> {
  met_dt[,gene:=<span class="hljs-string">&quot;NA&quot;</span>]
}

<span class="hljs-comment"># Accessibility</span>
<span class="hljs-keyword">if</span> (opts$overlapGenes) {
  acc_list &lt;- list()
  <span class="hljs-keyword">for</span> (i <span class="hljs-keyword">in</span> unique(acc_dt$anno)){

    <span class="hljs-comment"># Subset corresponding anno</span>
    acc_tmp &lt;- acc_dt[anno==i, ]

    <span class="hljs-comment"># Non gene-associated feature</span>
    <span class="hljs-keyword">if</span> (all(grepl(<span class="hljs-string">&quot;ENSMUSG&quot;</span>, unique(acc_tmp$id)) == <span class="hljs-literal">FALSE</span>)) {

      <span class="hljs-comment"># Extract coordiantes for methylation sites and for genes</span>
      feature_metadata_tmp &lt;- feature_metadata_filt.acc[anno==i, c(<span class="hljs-string">&quot;chr&quot;</span>,<span class="hljs-string">&quot;start&quot;</span>,<span class="hljs-string">&quot;end&quot;</span>,<span class="hljs-string">&quot;id&quot;</span>)] %&gt;% 
        .[,c(<span class="hljs-string">&quot;start&quot;</span>,<span class="hljs-string">&quot;end&quot;</span>) := list(start - opts$gene_window, end + opts$gene_window)] %&gt;% setkey(chr,start,end)

      <span class="hljs-comment"># Do the overlap</span>
      ov &lt;- foverlaps(
        gene_metadata_filt, 
        feature_metadata_tmp, 
        nomatch=<span class="hljs-number">0</span>) %&gt;% .[,c(<span class="hljs-string">&quot;gene&quot;</span>, <span class="hljs-string">&quot;id&quot;</span>)]

      <span class="hljs-comment"># If a feature overlaps with multiple genes, collapse them</span>
      ov1 &lt;- ov[is.na(gene)]
      ov2 &lt;- ov[!is.na(gene)] %&gt;% .[,.(gene=paste(gene,collapse=<span class="hljs-string">&quot;_&quot;</span>)), by=<span class="hljs-string">&quot;id&quot;</span>]
      ov &lt;- rbind(ov1,ov2)

      <span class="hljs-comment"># Merge with methylation data</span>
      acc_list[[i]] &lt;- merge(acc_tmp, ov, by=<span class="hljs-string">&quot;id&quot;</span>, allow.cartesian=<span class="hljs-literal">T</span>) 
    }
    <span class="hljs-comment"># Gene-associated feature</span>
    <span class="hljs-keyword">else</span> <span class="hljs-keyword">if</span> (all(grepl(<span class="hljs-string">&quot;ENSMUSG&quot;</span>, unique(acc_tmp$id)) == <span class="hljs-literal">TRUE</span>)) {
      acc_list[[i]] &lt;- merge(acc_tmp, gene_metadata[,c(<span class="hljs-string">&quot;id&quot;</span>,<span class="hljs-string">&quot;gene&quot;</span>)], by=<span class="hljs-string">&quot;id&quot;</span>)
    }
  }
  acc_dt &lt;- rbindlist(acc_list)
  rm(acc_list, acc_tmp,feature_metadata_tmp,ov)
} <span class="hljs-keyword">else</span> {
  acc_dt[,gene:=<span class="hljs-string">&quot;NA&quot;</span>]
}
</code></pre>
<h3 id="210-&#x6570;&#x636E;&#x8FC7;&#x6EE4;">2.10 &#x6570;&#x636E;&#x8FC7;&#x6EE4;</h3>
<pre><code class="lang-R"><span class="hljs-comment">#############################</span>
<span class="hljs-comment">## Filter methylation data ##</span>
<span class="hljs-comment">#############################</span>

<span class="hljs-comment"># Filter features by minimum number of CpGs</span>
met_dt &lt;- met_dt[N&gt;=opts$met_min.CpGs]

<span class="hljs-comment"># Filter features by  minimum number of cells</span>
met_dt[,N:=.N,by=c(<span class="hljs-string">&quot;id&quot;</span>,<span class="hljs-string">&quot;anno&quot;</span>,<span class="hljs-string">&quot;gene&quot;</span>)]  %&gt;% .[N&gt;=opts$met_min.cells] %&gt;% .[,N:=<span class="hljs-literal">NULL</span>]

<span class="hljs-comment"># Filter features by variance</span>
met_dt &lt;- met_dt[,var:=var(m), by=c(<span class="hljs-string">&quot;id&quot;</span>,<span class="hljs-string">&quot;anno&quot;</span>)] %&gt;% .[var&gt;<span class="hljs-number">0</span>] %&gt;% .[,var:=<span class="hljs-literal">NULL</span>] %&gt;% droplevels()

<span class="hljs-comment">###############################</span>
<span class="hljs-comment">## Filter accessibility data ##</span>
<span class="hljs-comment">###############################</span>

<span class="hljs-comment"># Filter features by minimum number of GpCs</span>
acc_dt &lt;- acc_dt[N&gt;=opts$acc_min.GpCs]

<span class="hljs-comment"># Filter features by  minimum number of cells</span>
acc_dt[,N:=.N,by=c(<span class="hljs-string">&quot;id&quot;</span>,<span class="hljs-string">&quot;anno&quot;</span>,<span class="hljs-string">&quot;gene&quot;</span>)]  %&gt;% .[N&gt;=opts$acc_min.cells] %&gt;% .[,N:=<span class="hljs-literal">NULL</span>]

<span class="hljs-comment"># Filter features by variance</span>
acc_dt &lt;- acc_dt[,var:=var(m), by=c(<span class="hljs-string">&quot;id&quot;</span>,<span class="hljs-string">&quot;anno&quot;</span>)] %&gt;% .[var&gt;<span class="hljs-number">0</span>] %&gt;% .[,var:=<span class="hljs-literal">NULL</span>] %&gt;% droplevels()

<span class="hljs-comment">################################</span>
<span class="hljs-comment">## Filter RNA expression data ##</span>
<span class="hljs-comment">################################</span>

<span class="hljs-comment"># Remove lowly expressed genes</span>
rna_dt &lt;- rna_dt[,mean:=mean(expr),by=<span class="hljs-string">&quot;ens_id&quot;</span>] %&gt;% .[mean&gt;<span class="hljs-number">0.1</span>] %&gt;% .[,mean:=<span class="hljs-literal">NULL</span>]

<span class="hljs-comment"># Remove genes with constant expression levels</span>
rna_dt &lt;- rna_dt[,var:=var(expr),by=<span class="hljs-string">&quot;ens_id&quot;</span>] %&gt;% .[var&gt;<span class="hljs-number">0.1</span>] %&gt;% .[,var:=<span class="hljs-literal">NULL</span>]

<span class="hljs-comment"># Filter genes with low cellular detection rate and sites with low coverage across samples</span>
rna_dt &lt;- rna_dt[,cdr:=sum(expr&gt;<span class="hljs-number">0</span>)/length(opts$rna_cells), by=<span class="hljs-string">&quot;ens_id&quot;</span>] %&gt;% .[cdr&gt;=opts$rna_min.cdr] %&gt;% .[,cdr:=<span class="hljs-literal">NULL</span>]

<span class="hljs-comment"># Extract top N highly variable genes</span>
rna_dt &lt;- rna_dt[,var:=var(expr), by=<span class="hljs-string">&quot;ens_id&quot;</span>] %&gt;% .[var&gt;<span class="hljs-number">0</span>] %&gt;% .[,var:=<span class="hljs-literal">NULL</span>] %&gt;% droplevels()
</code></pre>
<h3 id="211-&#x534F;&#x53D8;&#x91CF;&#x56DE;&#x5F52;">2.11 &#x534F;&#x53D8;&#x91CF;&#x56DE;&#x5F52;</h3>
<p>&#x4F5C;&#x8005;&#x4F7F;&#x7528;&#x4E86;&#x6A21;&#x578B;&#x62DF;&#x5408;&#x7684;&#x6B8B;&#x5DEE;&#x4F5C;&#x4E3A;&#x57FA;&#x56E0;&#x8868;&#x8FBE;&#x7684;&#x6807;&#x51C6;&#x5316;&#x5B9A;&#x91CF;&#x3002;</p>
<pre><code class="lang-R"><span class="hljs-comment"># RNA: number of expressed genes</span>
foo &lt;- data.table(id_rna=colnames(sce), covariate=sce$total_features_by_counts/nrow(sce))
rna_dt &lt;- rna_dt %&gt;% merge(foo, by=<span class="hljs-string">&quot;id_rna&quot;</span>) %&gt;%
  .[,expr:=lm(formula=expr~covariate)[[<span class="hljs-string">&quot;residuals&quot;</span>]], by=c(<span class="hljs-string">&quot;gene&quot;</span>)] %&gt;%
  .[,covariate:=<span class="hljs-literal">NULL</span>]

<span class="hljs-comment"># RNA: strong batch effect between E7.5 plates</span>
<span class="hljs-keyword">if</span> (any(sample_metadata$stage==<span class="hljs-string">&quot;E7.5&quot;</span>)) {
  foo &lt;- sample_metadata[stage==<span class="hljs-string">&quot;E7.5&quot;</span>,c(<span class="hljs-string">&quot;id_rna&quot;</span>,<span class="hljs-string">&quot;plate&quot;</span>)] %&gt;%
    .[,plate:=as.factor(grepl(<span class="hljs-string">&quot;PS_VE&quot;</span>,plate))]
  rna_dt &lt;- rbind(
    rna_dt[id_rna%<span class="hljs-keyword">in</span>%foo$id_rna] %&gt;% merge(foo, by=<span class="hljs-string">&quot;id_rna&quot;</span>) %&gt;%
      .[,expr:=lm(formula=expr~plate)[[<span class="hljs-string">&quot;residuals&quot;</span>]], by=c(<span class="hljs-string">&quot;gene&quot;</span>)] %&gt;% .[,c(<span class="hljs-string">&quot;plate&quot;</span>):=<span class="hljs-literal">NULL</span>],
    rna_dt[!id_rna%<span class="hljs-keyword">in</span>%foo$id_rna]
  )
}

<span class="hljs-comment"># RNA: strong batch effect between E6.5 plates</span>
<span class="hljs-keyword">if</span> (any(sample_metadata$stage==<span class="hljs-string">&quot;E6.5&quot;</span>)) {
  foo &lt;- sample_metadata[stage==<span class="hljs-string">&quot;E6.5&quot;</span> &amp; plate%<span class="hljs-keyword">in</span>%c(<span class="hljs-string">&quot;E6.5_late_Plate1&quot;</span>,<span class="hljs-string">&quot;E6.5_late_Plate2&quot;</span>),c(<span class="hljs-string">&quot;id_rna&quot;</span>,<span class="hljs-string">&quot;plate&quot;</span>)]
  rna_dt &lt;- rbind(
    rna_dt[id_rna%<span class="hljs-keyword">in</span>%foo$id_rna] %&gt;% merge(foo, by=<span class="hljs-string">&quot;id_rna&quot;</span>) %&gt;%
      .[,expr:=lm(formula=expr~plate)[[<span class="hljs-string">&quot;residuals&quot;</span>]], by=c(<span class="hljs-string">&quot;gene&quot;</span>)] %&gt;% .[,plate:=<span class="hljs-literal">NULL</span>],
    rna_dt[!id_rna%<span class="hljs-keyword">in</span>%foo$id_rna]
  )
}

<span class="hljs-comment"># Methylation: differences in mean methylation rate</span>
foo &lt;- met_dt[,.(covariate=mean(m)),by=c(<span class="hljs-string">&quot;id_met&quot;</span>)]
foo &lt;- fread(io$met.stats) %&gt;%
  .[,mean:=log2(((mean/<span class="hljs-number">100</span>)+<span class="hljs-number">0.01</span>)/(<span class="hljs-number">1</span>-(mean/<span class="hljs-number">100</span>)+<span class="hljs-number">0.01</span>))] %&gt;%
  .[,c(<span class="hljs-string">&quot;id_met&quot;</span>,<span class="hljs-string">&quot;mean&quot;</span>)]
met_dt &lt;- met_dt %&gt;% merge(foo, by=<span class="hljs-string">&quot;id_met&quot;</span>) %&gt;%
  .[,m:=mean(m) + lm(formula=m~mean)[[<span class="hljs-string">&quot;residuals&quot;</span>]], by=c(<span class="hljs-string">&quot;id&quot;</span>,<span class="hljs-string">&quot;anno&quot;</span>)]


<span class="hljs-comment"># Accessibility: differences in global accessibility rate</span>
foo &lt;- fread(io$acc.stats) %&gt;%
  .[,mean:=log2(((mean/<span class="hljs-number">100</span>)+<span class="hljs-number">0.01</span>)/(<span class="hljs-number">1</span>-(mean/<span class="hljs-number">100</span>)+<span class="hljs-number">0.01</span>))] %&gt;%
  .[,c(<span class="hljs-string">&quot;id_acc&quot;</span>,<span class="hljs-string">&quot;mean&quot;</span>)]
acc_dt &lt;- acc_dt %&gt;% merge(foo, by=<span class="hljs-string">&quot;id_acc&quot;</span>) %&gt;%
  .[,m:=mean(m) + lm(formula=m~mean)[[<span class="hljs-string">&quot;residuals&quot;</span>]], by=c(<span class="hljs-string">&quot;id&quot;</span>,<span class="hljs-string">&quot;anno&quot;</span>)]
</code></pre>
<h3 id="212-&#x7279;&#x5F81;&#x9009;&#x62E9;">2.12 &#x7279;&#x5F81;&#x9009;&#x62E9;</h3>
<p>&#x5728;&#x8FD9;&#x91CC;&#xFF0C;&#x4F5C;&#x8005;&#x9009;&#x62E9;&#x4E86;N&#x4E2A;&#x6700;&#x5927;&#x7684;&#x6837;&#x672C;&#x65B9;&#x5DEE;&#x6700;&#x5927;&#x7684;&#x503C;&#xFF0C;&#x6BD4;&#x5982;&#x9009;&#x62E9;&#x4E86;&#x524D;N&#x4E2A;&#x57FA;&#x56E0;&#x8868;&#x8FBE;&#x65B9;&#x5DEE;&#x6700;&#x5927;&#x7684;&#x57FA;&#x56E0;</p>
<pre><code class="lang-R"><span class="hljs-comment"># RNA: Extract top N highly variable genes</span>
keep_hv_genes &lt;- rna_dt[,.(var=var(expr)), by=<span class="hljs-string">&quot;ens_id&quot;</span>] %&gt;% .[var&gt;<span class="hljs-number">0</span>] %&gt;% setorder(-var) %&gt;% head(n=opts$rna_ngenes) %&gt;% .$ens_id
rna_dt &lt;- rna_dt[ens_id%<span class="hljs-keyword">in</span>%as.character(keep_hv_genes)] %&gt;% droplevels()

<span class="hljs-comment"># Accessibility: Extract top N most variable features</span>
keep_hv_sites &lt;- acc_dt %&gt;% split(.$anno) %&gt;% map(~ .[,.(var=var(m)), by=<span class="hljs-string">&quot;id&quot;</span>] %&gt;% .[var&gt;<span class="hljs-number">0</span>] %&gt;% setorder(-var) %&gt;% head(n=opts$acc_nfeatures) %&gt;% .$id)
acc_dt &lt;- acc_dt %&gt;% split(.$anno) %&gt;% map2(.,names(.), <span class="hljs-keyword">function</span>(x,y) x[id %<span class="hljs-keyword">in</span>% keep_hv_sites[[y]]]) %&gt;% rbindlist %&gt;% droplevels()

<span class="hljs-comment"># Methylation: Extract top N most variable features</span>
keep_hv_sites &lt;- met_dt %&gt;% split(.$anno) %&gt;% map(~ .[,.(var=var(m)), by=<span class="hljs-string">&quot;id&quot;</span>] %&gt;% .[var&gt;<span class="hljs-number">0</span>] %&gt;% setorder(-var) %&gt;% head(n=opts$met_nfeatures) %&gt;% .$id)
met_dt &lt;- met_dt %&gt;% split(.$anno) %&gt;% map2(.,names(.), <span class="hljs-keyword">function</span>(x,y) x[id %<span class="hljs-keyword">in</span>% keep_hv_sites[[y]]]) %&gt;% rbindlist %&gt;% droplevels()

met_dt[,var(m),by=c(<span class="hljs-string">&quot;id&quot;</span>,<span class="hljs-string">&quot;anno&quot;</span>)]
</code></pre>
<h3 id="213-&#x5408;&#x5E76;&#x6570;&#x636E;">2.13 &#x5408;&#x5E76;&#x6570;&#x636E;</h3>
<p>&#x5728;&#x8FD9;&#x91CC;&#xFF0C;&#x4F5C;&#x8005;&#x5C06;&#x4E09;&#x4E2A;&#x8868;&#x5408;&#x5E76;&#x6210;&#x4E86;&#x4E00;&#x4E2A;data.txt&#x7528;&#x4E8E;&#x6A21;&#x578B;&#x5206;&#x6790;&#xFF0C;&#x6211;&#x4EEC;&#x5728;&#x4E0B;&#x4E00;&#x6B65;&#x7684;MOFA&#x5EFA;&#x6A21;&#x4E2D;&#x4F1A;&#x7528;&#x5230;&#x8FD9;&#x4E2A;data.txt</p>
<pre><code class="lang-R">data1 &lt;- rna_dt %&gt;% .[,c(<span class="hljs-string">&quot;sample&quot;</span>,<span class="hljs-string">&quot;gene&quot;</span>,<span class="hljs-string">&quot;expr&quot;</span>)] %&gt;%  
  setnames(c(<span class="hljs-string">&quot;sample&quot;</span>,<span class="hljs-string">&quot;feature&quot;</span>,<span class="hljs-string">&quot;value&quot;</span>)) %&gt;% .[,c(<span class="hljs-string">&quot;feature_group&quot;</span>,<span class="hljs-string">&quot;sample_group&quot;</span>):=list(<span class="hljs-string">&quot;RNA&quot;</span>,<span class="hljs-string">&quot;gastrulation&quot;</span>)]
data2 &lt;- met_dt %&gt;% .[,c(<span class="hljs-string">&quot;sample&quot;</span>,<span class="hljs-string">&quot;id&quot;</span>,<span class="hljs-string">&quot;m&quot;</span>,<span class="hljs-string">&quot;anno&quot;</span>)] %&gt;%  
  setnames(c(<span class="hljs-string">&quot;sample&quot;</span>,<span class="hljs-string">&quot;feature&quot;</span>,<span class="hljs-string">&quot;value&quot;</span>,<span class="hljs-string">&quot;feature_group&quot;</span>)) %&gt;% .[,c(<span class="hljs-string">&quot;feature&quot;</span>,<span class="hljs-string">&quot;feature_group&quot;</span>,<span class="hljs-string">&quot;sample_group&quot;</span>):=list(paste0(<span class="hljs-string">&quot;met_&quot;</span>,feature), paste0(<span class="hljs-string">&quot;met_&quot;</span>,feature_group), <span class="hljs-string">&quot;gastrulation&quot;</span>)]
data3 &lt;- acc_dt %&gt;% .[,c(<span class="hljs-string">&quot;sample&quot;</span>,<span class="hljs-string">&quot;id&quot;</span>,<span class="hljs-string">&quot;m&quot;</span>,<span class="hljs-string">&quot;anno&quot;</span>)] %&gt;%  
  setnames(c(<span class="hljs-string">&quot;sample&quot;</span>,<span class="hljs-string">&quot;feature&quot;</span>,<span class="hljs-string">&quot;value&quot;</span>,<span class="hljs-string">&quot;feature_group&quot;</span>)) %&gt;% .[,c(<span class="hljs-string">&quot;feature&quot;</span>,<span class="hljs-string">&quot;feature_group&quot;</span>,<span class="hljs-string">&quot;sample_group&quot;</span>):=list(paste0(<span class="hljs-string">&quot;acc_&quot;</span>,feature), paste0(<span class="hljs-string">&quot;acc_&quot;</span>,feature_group), <span class="hljs-string">&quot;gastrulation&quot;</span>)]

data &lt;- rbind(data1,data2,data3)

outfile &lt;- paste0(io$outdir,<span class="hljs-string">&quot;/data.txt&quot;</span>)
fwrite(data, file=outfile, col.names=<span class="hljs-literal">T</span>, quote=<span class="hljs-literal">F</span>, sep=<span class="hljs-string">&quot;\t&quot;</span>)
</code></pre>
<p>&#x4F5C;&#x8005;&#x540E;&#x7EED;&#x7684;&#x4EE3;&#x7801;&#x53EF;&#x80FD;MOFA&#x5305;&#x8DD1;&#x5F97;&#x52A8;&#xFF0C;&#x7531;&#x4E8E;&#x6211;&#x662F;MOFA2&#x5305;&#xFF0C;&#x6240;&#x4EE5;&#x540E;&#x9762;&#x7684;&#x6211;&#x5C31;&#x8DD1;&#x4E0D;&#x52A8;&#x4E86;&#xFF0C;&#x5F00;&#x59CB;&#x5207;&#x73AF;&#x5883;&#x8BAD;&#x7EC3;&#x6A21;&#x578B;</p>
<h2 id="3&#x3001;mofa&#x6A21;&#x578B;&#x8BAD;&#x7EC3;">3&#x3001;MOFA&#x6A21;&#x578B;&#x8BAD;&#x7EC3;</h2>
<p>&#x8BE5;&#x6A21;&#x578B;&#x8BAD;&#x7EC3;&#x6211;&#x7528;&#x4E0A;&#x4E86;GPU&#xFF0C;&#x6240;&#x4EE5;&#x5207;&#x5230;&#x4E86;colab&#x4E0A;</p>
<h3 id="31-&#x4F9D;&#x8D56;&#x51C6;&#x5907;">3.1 &#x4F9D;&#x8D56;&#x51C6;&#x5907;</h3>
<p>mofapy2&#x8FD9;&#x4E2A;&#x5305;&#x6709;&#x70B9;&#x795E;&#x5947;&#xFF0C;&#x6211;&#x76F4;&#x63A5;&#x5728;jupyter&#x91CC;&#x88C5;&#x4F1A;&#x62A5;&#x9519;&#xFF0C;&#x6240;&#x4EE5;&#x6211;&#x6253;&#x5F00;&#x4E86;&#x7EC8;&#x7AEF;&#xFF0C;&#x8F93;&#x5165;</p>
<pre><code>pip install mofapy2
</code></pre><p>&#x5B89;&#x88C5;&#x5B8C;&#x540E;&#xFF0C;&#x6211;&#x4EEC;&#x5BFC;&#x5165;&#x5305;</p>
<pre><code class="lang-python"><span class="hljs-keyword">from</span> mofapy2.run.entry_point <span class="hljs-keyword">import</span> entry_point
<span class="hljs-keyword">import</span> pandas <span class="hljs-keyword">as</span> pd
<span class="hljs-keyword">import</span> numpy <span class="hljs-keyword">as</span> np

<span class="hljs-comment"># initialise the entry point</span>
ent = entry_point()
</code></pre>
<p>&#x51FA;&#x73B0;&#x4EE5;&#x4E0B;&#x56FE;&#x6848;&#x5C31;&#x662F;&#x6210;&#x529F;</p>
<pre><code class="lang-python">        <span class="hljs-comment">#########################################################</span>
        <span class="hljs-comment">###           __  __  ____  ______                    ### </span>
        <span class="hljs-comment">###          |  \/  |/ __ \|  ____/\    _             ### </span>
        <span class="hljs-comment">###          | \  / | |  | | |__ /  \ _| |_           ### </span>
        <span class="hljs-comment">###          | |\/| | |  | |  __/ /\ \_   _|          ###</span>
        <span class="hljs-comment">###          | |  | | |__| | | / ____ \|_|            ###</span>
        <span class="hljs-comment">###          |_|  |_|\____/|_|/_/    \_\              ###</span>
        <span class="hljs-comment">###                                                   ### </span>
        <span class="hljs-comment">#########################################################</span>
</code></pre>
<h3 id="32-&#x6570;&#x636E;&#x6E05;&#x6D17;">3.2 &#x6570;&#x636E;&#x6E05;&#x6D17;</h3>
<p>&#x6211;&#x4EEC;&#x524D;&#x9762;&#x5F97;&#x5230;&#x4E86;data.txt&#x8FD9;&#x4E2A;&#x6570;&#x636E;&#x6587;&#x4EF6;&#xFF0C;&#x4E8E;&#x662F;&#x6211;&#x4EEC;&#x9700;&#x8981;&#x7A0D;&#x5FAE;&#x6539;&#x4E00;&#x4E0B;columns&#xFF0C;&#x4F7F;&#x5176;&#x7B26;&#x5408;mofapy2&#x7684;&#x8981;&#x6C42;</p>
<pre><code class="lang-python">data=pd.read_csv(<span class="hljs-string">&apos;/content/drive/MyDrive/data/data.txt&apos;</span>,sep=<span class="hljs-string">&apos;\t&apos;</span>)
data.columns=[<span class="hljs-string">&apos;sample&apos;</span>,<span class="hljs-string">&apos;feature&apos;</span>,<span class="hljs-string">&apos;value&apos;</span>,<span class="hljs-string">&apos;view&apos;</span>,<span class="hljs-string">&apos;group&apos;</span>]
data.head()
</code></pre>
<p>&#x6210;&#x529F;&#x8FD4;&#x56DE;&#x4E86;&#x5982;&#x4E0B;&#x5185;&#x5BB9;</p>
<pre><code class="lang-python">    sample    feature    value    view    group
<span class="hljs-number">0</span>    E6<span class="hljs-number">.75</span>_Plate1_E1    Cdc45    <span class="hljs-number">-2.005848</span>    RNA    gastrulation
<span class="hljs-number">1</span>    E6<span class="hljs-number">.75</span>_Plate1_E1    Narf    <span class="hljs-number">-1.188268</span>    RNA    gastrulation
<span class="hljs-number">2</span>    E6<span class="hljs-number">.75</span>_Plate1_E1    Klf6    <span class="hljs-number">4.435413</span>    RNA    gastrulation
<span class="hljs-number">3</span>    E6<span class="hljs-number">.75</span>_Plate1_E1    Wnt3    <span class="hljs-number">-0.727745</span>    RNA    gastrulation
<span class="hljs-number">4</span>    E6<span class="hljs-number">.75</span>_Plate1_E1    Fer    <span class="hljs-number">-0.995191</span>    RNA    gastrulation
</code></pre>
<h3 id="33-&#x6A21;&#x578B;&#x6570;&#x636E;&#x8BBE;&#x7F6E;">3.3 &#x6A21;&#x578B;&#x6570;&#x636E;&#x8BBE;&#x7F6E;</h3>
<ul>
<li><strong>scale_groups</strong>: if groups have different ranges/variances, it is good practice to scale each group to unit variance. Default is False</li>
<li><strong>scale_views</strong>: if views have different ranges/variances, it is good practice to scale each view to unit variance. Default is False</li>
</ul>
<pre><code class="lang-python">ent.set_data_options(
    scale_groups = <span class="hljs-keyword">False</span>, 
    scale_views = <span class="hljs-keyword">False</span>
)
</code></pre>
<h3 id="34-&#x5BFC;&#x5165;&#x6570;&#x636E;&#x5230;&#x6A21;&#x578B;&#x4E2D;">3.4 &#x5BFC;&#x5165;&#x6570;&#x636E;&#x5230;&#x6A21;&#x578B;&#x4E2D;</h3>
<ul>
<li><strong>likelihoods</strong>: a list of strings, either &quot;gaussian&quot;, &quot;poisson&quot; or &quot;bernoulli&quot;. If None (default), they are guessed internally</li>
</ul>
<p>&#x5728;&#x8FD9;&#x91CC;&#xFF0C;&#x6211;&#x4EEC;&#x8BA4;&#x4E3A;&#x6BCF;&#x4E00;&#x4E2A;&#x7EC4;&#x5B66;&#x6570;&#x636E;&#x90FD;&#x662F;gaussian&#x5206;&#x5E03;</p>
<pre><code class="lang-python">ent.set_data_df(data, likelihoods = [<span class="hljs-string">&quot;gaussian&quot;</span>,<span class="hljs-string">&quot;gaussian&quot;</span>,<span class="hljs-string">&quot;gaussian&quot;</span>,<span class="hljs-string">&quot;gaussian&quot;</span>,<span class="hljs-string">&quot;gaussian&quot;</span>,<span class="hljs-string">&quot;gaussian&quot;</span>,<span class="hljs-string">&quot;gaussian&quot;</span>,<span class="hljs-string">&quot;gaussian&quot;</span>,<span class="hljs-string">&quot;gaussian&quot;</span>])
</code></pre>
<p>&#x7A0B;&#x5E8F;&#x8F93;&#x51FA;&#x7ED3;&#x679C;&#xFF1A;</p>
<pre><code>Loaded group=&apos;gastrulation&apos; view=&apos;RNA&apos; with N=2147 samples and D=2500 features...
Loaded group=&apos;gastrulation&apos; view=&apos;acc_H3K27ac_distal_E7.5_Ect_intersect12&apos; with N=729 samples and D=1000 features...
Loaded group=&apos;gastrulation&apos; view=&apos;acc_H3K27ac_distal_E7.5_End_intersect12&apos; with N=729 samples and D=1000 features...
Loaded group=&apos;gastrulation&apos; view=&apos;acc_H3K27ac_distal_E7.5_Mes_intersect12&apos; with N=729 samples and D=1000 features...
Loaded group=&apos;gastrulation&apos; view=&apos;acc_prom_2000_2000&apos; with N=729 samples and D=1000 features...
Loaded group=&apos;gastrulation&apos; view=&apos;met_H3K27ac_distal_E7.5_Ect_intersect12&apos; with N=826 samples and D=1000 features...
Loaded group=&apos;gastrulation&apos; view=&apos;met_H3K27ac_distal_E7.5_End_intersect12&apos; with N=826 samples and D=1000 features...
Loaded group=&apos;gastrulation&apos; view=&apos;met_H3K27ac_distal_E7.5_Mes_intersect12&apos; with N=826 samples and D=1000 features...
Loaded group=&apos;gastrulation&apos; view=&apos;met_prom_2000_2000&apos; with N=826 samples and D=1000 features...
</code></pre><h3 id="35-&#x6A21;&#x578B;&#x53C2;&#x6570;&#x8BBE;&#x7F6E;">3.5 &#x6A21;&#x578B;&#x53C2;&#x6570;&#x8BBE;&#x7F6E;</h3>
<ul>
<li><strong>factors</strong>: number of factors</li>
<li><strong>spikeslab_weights</strong>: use spike-slab sparsity prior in the weights? default is TRUE</li>
<li><strong>ard_factors</strong>: use ARD prior in the factors? Default is TRUE if using multiple groups. This is guessed by default.</li>
<li><strong>ard_weights</strong>: use ARD prior in the weights? Default is TRUE if using multiple views. This is guessed by default.</li>
</ul>
<pre><code class="lang-python">ent.set_model_options(
    factors = <span class="hljs-number">10</span>, 
    spikeslab_weights = <span class="hljs-keyword">True</span>, 
    ard_factors = <span class="hljs-keyword">True</span>,
    ard_weights = <span class="hljs-keyword">True</span>
)
</code></pre>
<p>&#x7A0B;&#x5E8F;&#x8F93;&#x51FA;&#x7ED3;&#x679C;</p>
<pre><code>Model options:
- Automatic Relevance Determination prior on the factors: True
- Automatic Relevance Determination prior on the weights: True
- Spike-and-slab prior on the factors: False
- Spike-and-slab prior on the weights: True
Likelihoods:
- View 0 (RNA): gaussian
- View 1 (acc_H3K27ac_distal_E7.5_Ect_intersect12): gaussian
- View 2 (acc_H3K27ac_distal_E7.5_End_intersect12): gaussian
- View 3 (acc_H3K27ac_distal_E7.5_Mes_intersect12): gaussian
- View 4 (acc_prom_2000_2000): gaussian
- View 5 (met_H3K27ac_distal_E7.5_Ect_intersect12): gaussian
- View 6 (met_H3K27ac_distal_E7.5_End_intersect12): gaussian
- View 7 (met_H3K27ac_distal_E7.5_Mes_intersect12): gaussian
- View 8 (met_prom_2000_2000): gaussian
</code></pre><h3 id="36-&#x6A21;&#x578B;&#x8BAD;&#x7EC3;&#x8BBE;&#x7F6E;">3.6 &#x6A21;&#x578B;&#x8BAD;&#x7EC3;&#x8BBE;&#x7F6E;</h3>
<ul>
<li><strong>iter</strong>: number of iterations. Default is 1000.</li>
<li><strong>convergence_mode</strong>: &quot;fast&quot;, &quot;medium&quot;, &quot;slow&quot;. For exploration, the fast mode is good enough.</li>
<li><strong>startELBO</strong>: initial iteration to compute the ELBO (the objective function used to assess convergence)</li>
<li><strong>freqELBO</strong>: frequency of computations of the ELBO (the objective function used to assess convergence)</li>
<li><strong>dropR2</strong>: minimum variance explained criteria to drop factors while training gpu_mode: use GPU mode? (needs cupy installed and a functional GPU, see <a href="https://cupy.chainer.org/" target="_blank">https://cupy.chainer.org/</a>)</li>
<li><strong>verbose</strong>: verbose mode?</li>
<li><strong>seed</strong>: random seed</li>
</ul>
<pre><code class="lang-python">ent.set_train_options(
    iter = <span class="hljs-number">5000</span>, 
    convergence_mode = <span class="hljs-string">&quot;slow&quot;</span>, 
    startELBO = <span class="hljs-number">1</span>, 
    freqELBO = <span class="hljs-number">1</span>, 
    dropR2 = <span class="hljs-number">0.001</span>, 
    gpu_mode = <span class="hljs-keyword">True</span>, 
    verbose = <span class="hljs-keyword">False</span>, 
    seed = <span class="hljs-number">42</span>
)
</code></pre>
<h3 id="37-&#xFF08;&#x53EF;&#x9009;&#xFF09;&#x968F;&#x673A;&#x63A8;&#x7406;&#x9009;&#x9879;">3.7 &#xFF08;&#x53EF;&#x9009;&#xFF09;&#x968F;&#x673A;&#x63A8;&#x7406;&#x9009;&#x9879;</h3>
<p>&#x5982;&#x679C;&#x6837;&#x672C;&#x6570;&#x91CF;&#x975E;&#x5E38;&#x5927;&#xFF08;&#x4EE5;&gt;1e4 &#x7684;&#x987A;&#x5E8F;&#xFF09;&#xFF0C;&#x60A8;&#x53EF;&#x80FD;&#x60F3;&#x5C1D;&#x8BD5;&#x968F;&#x673A;&#x63A8;&#x7406;&#x65B9;&#x6848;&#x3002; &#x7136;&#x800C;&#xFF0C;&#x5B83;&#x9700;&#x8981;&#x4E00;&#x4E9B;&#x989D;&#x5916;&#x7684;&#x8D85;&#x53C2;&#x6570;&#xFF0C;&#x5728;&#x67D0;&#x4E9B;&#x6570;&#x636E;&#x96C6;&#x4E2D;&#x53EF;&#x80FD;&#x9700;&#x8981;&#x4F18;&#x5316;&#xFF08;&#x53C2;&#x89C1;&#x968F;&#x673A;&#x5C0F;&#x63D2;&#x56FE;&#xFF09;&#xFF0C;&#x5B9E;&#x9645;&#x4E0A;&#x90FD;&#x662F;&#x673A;&#x5668;&#x5B66;&#x4E60;&#x7684;&#x53C2;&#x6570;</p>
<ul>
<li><strong>batch_size</strong>: float value indicating the batch size (as a fraction of the total data set: either 0.10, 0.25 or 0.50)</li>
<li><strong>learning_rate</strong>: learning rate (we recommend values from 0.25 to 0.50)</li>
<li><strong>forgetting_rate</strong>: forgetting rate (we recommend values from 0.75 to 1.0)</li>
</ul>
<pre><code class="lang-python"><span class="hljs-comment"># We do not want to use stochastic inference for this data set</span>

<span class="hljs-comment"># ent.set_stochastic_options(</span>
<span class="hljs-comment">#     batch_size = 0.5,</span>
<span class="hljs-comment">#     learning_rate = 0.5, </span>
<span class="hljs-comment">#     forgetting_rate = 0.25</span>
<span class="hljs-comment"># )</span>
</code></pre>
<h3 id="38-&#x8BAD;&#x7EC3;mofa&#x6A21;&#x578B;">3.8 &#x8BAD;&#x7EC3;MOFA&#x6A21;&#x578B;</h3>
<p>&#x7EC8;&#x4E8E;&#x5230;&#x8FD9;&#x4E00;&#x6B65;&#x4E86;&#xFF0C;&#x8FD8;&#x662F;&#x6BD4;&#x8F83;&#x7B80;&#x5355;&#x7684;</p>
<pre><code class="lang-python">ent.build()

ent.run()
</code></pre>
<pre><code>######################################
## Training the model with seed 42 ##
######################################


ELBO before training: -80858378.90 

Iteration 1: time=8.77, ELBO=-17216582.96, deltaELBO=63641795.939 (78.70773173%), Factors=9
Iteration 2: time=3.07, ELBO=-17158861.17, deltaELBO=57721.785 (0.07138627%), Factors=9
Iteration 3: time=3.65, ELBO=-17147286.77, deltaELBO=11574.401 (0.01431441%), Factors=9
......
Iteration 156: time=3.25, ELBO=-17055635.59, deltaELBO=4.499 (0.00000556%), Factors=9
Iteration 157: time=3.22, ELBO=-17055631.25, deltaELBO=4.337 (0.00000536%), Factors=9
Iteration 158: time=3.23, ELBO=-17055627.07, deltaELBO=4.189 (0.00000518%), Factors=9
Iteration 159: time=3.34, ELBO=-17055623.01, deltaELBO=4.053 (0.00000501%), Factors=9
Iteration 160: time=3.34, ELBO=-17055619.08, deltaELBO=3.930 (0.00000486%), Factors=9
Iteration 161: time=3.36, ELBO=-17055615.26, deltaELBO=3.819 (0.00000472%), Factors=9

Converged!



#######################
## Training finished ##
#######################
</code></pre><h3 id="39-&#x6A21;&#x578B;&#x7684;&#x4FDD;&#x5B58;">3.9 &#x6A21;&#x578B;&#x7684;&#x4FDD;&#x5B58;</h3>
<pre><code>ent.save(outfile=&apos;model_1.hdf5&apos;)
</code></pre><p>&#x5230;&#x6B64;&#xFF0C;&#x6211;&#x4EEC;&#x6210;&#x529F;&#x901A;&#x8FC7;&#x591A;&#x7EC4;&#x5B66;&#x6570;&#x636E;&#x6210;&#x529F;&#x8BAD;&#x7EC3;&#x51FA;&#x4E86;&#x4E00;&#x4E2A;MOFA&#x6A21;&#x578B;&#xFF0C;&#x4E0D;&#x8FC7;&#x8FDB;&#x4E00;&#x6B65;&#x7684;&#x5206;&#x6790;&#x8981;&#x5230;R&#x8BED;&#x8A00;&#x4E2D;&#x8FDB;&#x884C;&#x4E86;</p>

                                
                                </section>
                            
    </div>
    <div class="search-results">
        <div class="has-results">
            
            <h1 class="search-results-title"><span class='search-results-count'></span> results matching "<span class='search-query'></span>"</h1>
            <ul class="search-results-list"></ul>
            
        </div>
        <div class="no-results">
            
            <h1 class="search-results-title">No results matching "<span class='search-query'></span>"</h1>
            
        </div>
    </div>
</div>

                        </div>
                    </div>
                
            </div>

            
                
                <a href="gastrulation_1.html" class="navigation navigation-prev " aria-label="Previous page: 1-1. 环境配置">
                    <i class="fa fa-angle-left"></i>
                </a>
                
                
                <a href="gastrulation_3.html" class="navigation navigation-next " aria-label="Next page: 1-3. 下游分析-Figure3b">
                    <i class="fa fa-angle-right"></i>
                </a>
                
            
        
    </div>

    <script>
        var gitbook = gitbook || [];
        gitbook.push(function() {
            gitbook.page.hasChanged({"page":{"title":"1-2. MOFA模型-Figure3b","level":"1.3.2","depth":2,"next":{"title":"1-3. 下游分析-Figure3b","level":"1.3.3","depth":2,"path":"GASTRULATION/gastrulation_3.md","ref":"GASTRULATION/gastrulation_3.md","articles":[]},"previous":{"title":"1-1. 环境配置","level":"1.3.1","depth":2,"path":"GASTRULATION/gastrulation_1.md","ref":"GASTRULATION/gastrulation_1.md","articles":[]},"dir":"ltr"},"config":{"gitbook":"*","theme":"default","variables":{},"plugins":["expandable-chapters"],"pluginsConfig":{"expandable-chapters":{},"highlight":{},"search":{},"lunr":{"maxIndexSize":1000000,"ignoreSpecialCharacters":false},"sharing":{"facebook":true,"twitter":true,"google":false,"weibo":false,"instapaper":false,"vk":false,"all":["facebook","google","twitter","weibo","instapaper"]},"fontsettings":{"theme":"white","family":"sans","size":2},"theme-default":{"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"showLevel":false}},"structure":{"langs":"LANGS.md","readme":"README.md","glossary":"GLOSSARY.md","summary":"SUMMARY.md"},"pdf":{"pageNumbers":true,"fontSize":12,"fontFamily":"Arial","paperSize":"a4","chapterMark":"pagebreak","pageBreaksBefore":"/","margin":{"right":62,"left":62,"top":56,"bottom":56}},"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"}},"file":{"path":"GASTRULATION/gastrulation_2.md","mtime":"2021-10-06T20:11:56.383Z","type":"markdown"},"gitbook":{"version":"3.2.3","time":"2021-10-08T15:50:43.701Z"},"basePath":"..","book":{"language":""}});
        });
    </script>
</div>

        
    <script src="../gitbook/gitbook.js"></script>
    <script src="../gitbook/theme.js"></script>
    
        
        <script src="../gitbook/gitbook-plugin-expandable-chapters/expandable-chapters.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search-engine.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/lunr.min.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/search-lunr.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-sharing/buttons.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-fontsettings/fontsettings.js"></script>
        
    

    </body>
</html>

