
<!DOCTYPE HTML>
<html lang="" >
    <head>
        <meta charset="UTF-8">
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <title>1-3. 下游分析-Figure3b · GitBook</title>
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta name="description" content="">
        <meta name="generator" content="GitBook 3.2.3">
        
        
        
    
    <link rel="stylesheet" href="../gitbook/style.css">

    
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-expandable-chapters/expandable-chapters.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-highlight/website.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-search/search.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-fontsettings/website.css">
                
            
        

    

    
        
    
        
    
        
    
        
    
        
    
        
    

        
    
    
    <meta name="HandheldFriendly" content="true"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../gitbook/images/apple-touch-icon-precomposed-152.png">
    <link rel="shortcut icon" href="../gitbook/images/favicon.ico" type="image/x-icon">

    
    
    <link rel="prev" href="gastrulation_2.html" />
    

    </head>
    <body>
        
<div class="book">
    <div class="book-summary">
        
            
<div id="book-search-input" role="search">
    <input type="text" placeholder="Type to search" />
</div>

            
                <nav role="navigation">
                


<ul class="summary">
    
    

    

    
        
        
    
        <li class="chapter " data-level="1.1" data-path="../">
            
                <a href="../">
            
                    
                    目录
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2" data-path="./">
            
                <a href="./">
            
                    
                    论文1: 单细胞分辨率下小鼠原肠胚形成的多组学分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.2.1" data-path="gastrulation_1.html">
            
                <a href="gastrulation_1.html">
            
                    
                    1-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.2" data-path="gastrulation_2.html">
            
                <a href="gastrulation_2.html">
            
                    
                    1-2. MOFA模型-Figure3b
            
                </a>
            

            
        </li>
    
        <li class="chapter active" data-level="1.2.3" data-path="gastrulation_3.html">
            
                <a href="gastrulation_3.html">
            
                    
                    1-3. 下游分析-Figure3b
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    

    

    <li class="divider"></li>

    <li>
        <a href="https://www.gitbook.com" target="blank" class="gitbook-link">
            Published with GitBook
        </a>
    </li>
</ul>


                </nav>
            
        
    </div>

    <div class="book-body">
        
            <div class="body-inner">
                
                    

<div class="book-header" role="navigation">
    

    <!-- Title -->
    <h1>
        <i class="fa fa-circle-o-notch fa-spin"></i>
        <a href=".." >1-3. 下游分析-Figure3b</a>
    </h1>
</div>




                    <div class="page-wrapper" tabindex="-1" role="main">
                        <div class="page-inner">
                            
<div id="book-search-results">
    <div class="search-noresults">
    
                                <section class="normal markdown-section">
                                
                                <h1 id="nature&#x590D;&#x73B0;&#xFF1A;&#x4E0B;&#x6E38;&#x5206;&#x6790;-figure3b">Nature&#x590D;&#x73B0;&#xFF1A;&#x4E0B;&#x6E38;&#x5206;&#x6790;-Figure3b</h1>
<p>&#x5728;&#x524D;&#x9762;&#x7684;&#x6559;&#x7A0B;&#x91CC;&#xFF0C;&#x6211;&#x4EEC;&#x5DF2;&#x7ECF;&#x5F97;&#x5230;&#x4E86;&#x4E00;&#x4E2A;MOFA&#x6A21;&#x578B;&#xFF0C;&#x5F88;&#x591A;&#x5206;&#x6790;&#x90FD;&#x53EF;&#x4EE5;&#x5F00;&#x59CB;&#x8FDB;&#x884C;&#x4E86;&#xFF0C;&#x53EA;&#x4E0D;&#x8FC7;&#x4E0B;&#x9762;&#x7684;&#x6559;&#x7A0B;&#x53C8;&#x5C06;&#x56DE;&#x5230;R&#x8BED;&#x8A00;&#x4E2D;&#x8FDB;&#x884C;&#xFF0C;&#x5728;&#x8FD9;&#x91CC;&#xFF0C;&#x6211;&#x6CA1;&#x7528;wsl&#x4E0A;&#x9762;&#x7684;R&#xFF0C;&#x800C;&#x662F;&#x672C;&#x5730;&#x7684;R&#x73AF;&#x5883;</p>
<h2 id="1&#x3001;&#x591A;&#x7EC4;&#x5B66;&#x56E0;&#x5B50;&#x5206;&#x6790;">1&#x3001;&#x591A;&#x7EC4;&#x5B66;&#x56E0;&#x5B50;&#x5206;&#x6790;</h2>
<h3 id="11-&#x52A0;&#x8F7D;&#x4F9D;&#x8D56;">1.1 &#x52A0;&#x8F7D;&#x4F9D;&#x8D56;</h3>
<pre><code class="lang-R"><span class="hljs-keyword">library</span>(MOFA2)
<span class="hljs-keyword">library</span>(data.table)
<span class="hljs-keyword">library</span>(purrr)
<span class="hljs-keyword">library</span>(ggplot2)
<span class="hljs-keyword">library</span>(RColorBrewer)
</code></pre>
<h3 id="12-&#x57FA;&#x7840;&#x8BBE;&#x7F6E;">1.2 &#x57FA;&#x7840;&#x8BBE;&#x7F6E;</h3>
<pre><code class="lang-R"><span class="hljs-comment">## Define I/O ##</span>

io &lt;- list()
io$basedir &lt;- <span class="hljs-string">&quot;D:/drive/scnmt&quot;</span>
io$gene_metadata &lt;- <span class="hljs-string">&quot;D:/drive/scnmt/gene/Mmusculus_genes_BioMart.87.txt&quot;</span>
io$outdir &lt;- <span class="hljs-string">&quot;D:/drive/scnmt/result&quot;</span>

io$sample.metadata &lt;- paste0(io$basedir,<span class="hljs-string">&quot;/sample_metadata.txt&quot;</span>)
io$met.dir &lt;- paste0(io$basedir,<span class="hljs-string">&quot;/met/feature_level&quot;</span>)
io$acc.dir &lt;- paste0(io$basedir,<span class="hljs-string">&quot;/acc/feature_level&quot;</span>)
io$rna.file &lt;- paste0(io$basedir,<span class="hljs-string">&quot;/rna/SingleCellExperiment.rds&quot;</span>)
io$annos_dir  &lt;- paste0(io$basedir, <span class="hljs-string">&quot;/features/genomic_contexts&quot;</span>)

<span class="hljs-comment"># Folders with the global statistics per cell</span>
io$met.stats &lt;- paste0(io$basedir,<span class="hljs-string">&quot;/met/results/stats/sample_stats.txt&quot;</span>)
io$acc.stats &lt;- paste0(io$basedir,<span class="hljs-string">&quot;/acc/results/stats/sample_stats.txt&quot;</span>)

<span class="hljs-comment">## Define options ##</span>
opts &lt;- list()

<span class="hljs-comment"># Define which annotations to look at</span>
opts$met.annos &lt;- c(
  <span class="hljs-comment"># &quot;genebody&quot;,</span>
  <span class="hljs-string">&quot;prom_2000_2000&quot;</span>,
  <span class="hljs-string">&quot;H3K27ac_distal_E7.5_Mes_intersect12&quot;</span>,
  <span class="hljs-string">&quot;H3K27ac_distal_E7.5_Ect_intersect12&quot;</span>,
  <span class="hljs-string">&quot;H3K27ac_distal_E7.5_End_intersect12&quot;</span>
  <span class="hljs-comment"># &quot;H3K4me3_E7.5_Mes&quot;,</span>
  <span class="hljs-comment"># &quot;H3K4me3_E7.5_End&quot;,</span>
  <span class="hljs-comment"># &quot;H3K4me3_E7.5_Ect&quot;,</span>
)

opts$acc.annos &lt;- c(
  <span class="hljs-comment"># &quot;genebody&quot;,</span>
  <span class="hljs-string">&quot;prom_2000_2000&quot;</span>,
  <span class="hljs-string">&quot;H3K27ac_distal_E7.5_Mes_intersect12&quot;</span>,
  <span class="hljs-string">&quot;H3K27ac_distal_E7.5_Ect_intersect12&quot;</span>,
  <span class="hljs-string">&quot;H3K27ac_distal_E7.5_End_intersect12&quot;</span>
  <span class="hljs-comment"># &quot;H3K4me3_E7.5_Mes&quot;,</span>
  <span class="hljs-comment"># &quot;H3K4me3_E7.5_End&quot;,</span>
  <span class="hljs-comment"># &quot;H3K4me3_E7.5_Ect&quot;,</span>
)


<span class="hljs-comment"># Define which stage and lineages to look at </span>
opts$stage_lineage &lt;- c(

  <span class="hljs-comment"># E4.5</span>
  <span class="hljs-string">&quot;E4.5_Epiblast&quot;</span>,

  <span class="hljs-comment"># E5.5</span>
  <span class="hljs-string">&quot;E5.5_Epiblast&quot;</span>,

  <span class="hljs-comment"># E6.5</span>
  <span class="hljs-string">&quot;E6.5_Epiblast&quot;</span>,
  <span class="hljs-string">&quot;E6.5_Primitive_Streak&quot;</span>,
  <span class="hljs-string">&quot;E6.5_Mesoderm&quot;</span>,

  <span class="hljs-comment"># E7.5</span>
  <span class="hljs-string">&quot;E7.5_Epiblast&quot;</span>,
  <span class="hljs-string">&quot;E7.5_Primitive_Streak&quot;</span>,
  <span class="hljs-string">&quot;E7.5_Ectoderm&quot;</span>,
  <span class="hljs-string">&quot;E7.5_Endoderm&quot;</span>,
  <span class="hljs-string">&quot;E7.5_Mesoderm&quot;</span>
)

<span class="hljs-comment"># Filtering options for methylation</span>
opts$met_min.CpGs &lt;- <span class="hljs-number">1</span>        <span class="hljs-comment"># minimum number of CpG sites per feature</span>
opts$met_min.cells &lt;- <span class="hljs-number">25</span>      <span class="hljs-comment"># minimum number of cells per feature</span>
opts$met_nfeatures &lt;- <span class="hljs-number">1000</span>    <span class="hljs-comment"># maximum number of features per view (filter based on variance)</span>

<span class="hljs-comment"># Filtering options for accessibility</span>
opts$acc_min.GpCs &lt;- <span class="hljs-number">5</span>        <span class="hljs-comment"># minimum number of GpC sites per feature</span>
opts$acc_min.cells &lt;- <span class="hljs-number">25</span>      <span class="hljs-comment"># minimum number of cells per feature</span>
opts$acc_nfeatures &lt;- <span class="hljs-number">1000</span>    <span class="hljs-comment"># maximum number of features per view (filter based on variance)</span>

<span class="hljs-comment"># Filtering options for RNA</span>
opts$rna_min.cdr &lt;- <span class="hljs-number">0.25</span>      <span class="hljs-comment"># Remove genes with cellular detection rate smaller than opts$min.cdr</span>
opts$rna_ngenes &lt;- <span class="hljs-number">2500</span>       <span class="hljs-comment"># maximum number of genes (filter based on variance)</span>

<span class="hljs-comment"># Define colors</span>
opts$colors &lt;- c(
  <span class="hljs-string">&quot;Epiblast&quot;</span>=<span class="hljs-string">&quot;grey70&quot;</span>,
  <span class="hljs-string">&quot;Mesoderm&quot;</span>=<span class="hljs-string">&quot;#CD3278&quot;</span>,
  <span class="hljs-string">&quot;Primitive Streak&quot;</span>=<span class="hljs-string">&quot;sandybrown&quot;</span>,
  <span class="hljs-string">&quot;Endoderm&quot;</span>=<span class="hljs-string">&quot;#43CD80&quot;</span>,
  <span class="hljs-string">&quot;Ectoderm&quot;</span>=<span class="hljs-string">&quot;steelblue&quot;</span>
)

<span class="hljs-comment"># window length for the overlap between genes and features</span>
opts$overlapGenes  &lt;- <span class="hljs-literal">FALSE</span>
opts$gene_window  &lt;- <span class="hljs-number">5e4</span>

<span class="hljs-comment"># Define which cells to use</span>
tmp &lt;- fread(io$sample.metadata) %&gt;%
  .[,stage_lineage:=as.factor(paste(stage,lineage10x_2,sep=<span class="hljs-string">&quot;_&quot;</span>))] %&gt;%
  .[stage_lineage%<span class="hljs-keyword">in</span>%opts$stage_lineage]
opts$met_cells &lt;- tmp %&gt;% .[pass_metQC==<span class="hljs-literal">T</span>, id_met]
opts$rna_cells &lt;- tmp %&gt;% .[pass_rnaQC==<span class="hljs-literal">T</span>, id_rna]
opts$acc_cells &lt;- tmp %&gt;% .[pass_accQC==<span class="hljs-literal">T</span>, id_acc]


<span class="hljs-comment">##########################</span>
<span class="hljs-comment">## Load sample metadata ##</span>
<span class="hljs-comment">##########################</span>

sample_metadata &lt;- fread(io$sample.metadata) %&gt;%
  .[,stage_lineage:=paste(stage,lineage10x_2,sep=<span class="hljs-string">&quot;_&quot;</span>)] %&gt;%
  .[id_met%<span class="hljs-keyword">in</span>%opts$met_cells | id_rna %<span class="hljs-keyword">in</span>% opts$rna_cells | id_acc %<span class="hljs-keyword">in</span>% opts$acc_cells ]
</code></pre>
<h3 id="13-&#x6A21;&#x578B;&#x52A0;&#x8F7D;">1.3 &#x6A21;&#x578B;&#x52A0;&#x8F7D;</h3>
<p>&#x6211;&#x4EEC;&#x5C06;&#x4E0A;&#x9762;&#x751F;&#x6210;&#x7684;model_1&#x6A21;&#x578B;&#x653E;&#x5230;&#x4E86;hdf5&#x76EE;&#x5F55;&#x4E0B;</p>
<pre><code class="lang-R">model &lt;- load_model(paste0(io$outdir,<span class="hljs-string">&quot;/hdf5/model_1.hdf5&quot;</span>))
model
</code></pre>
<p>&#x67E5;&#x770B;&#x4E00;&#x4E0B;&#x6A21;&#x578B;&#x4FE1;&#x606F;</p>
<pre><code class="lang-R">Trained MOFA with the following characteristics: 
 Number of views: <span class="hljs-number">9</span> 
 Views names: RNA Ectoderm enhancers (acc) Endoderm enhancers (acc) Mesoderm enhancers (acc) Acc CGI Promoters Ectoderm enhancers (met) Endoderm enhancers (met) Mesoderm enhancers (met) Met Promoters 
 Number of features (per view): <span class="hljs-number">2500</span> <span class="hljs-number">1000</span> <span class="hljs-number">1000</span> <span class="hljs-number">1000</span> <span class="hljs-number">1000</span> <span class="hljs-number">1000</span> <span class="hljs-number">1000</span> <span class="hljs-number">1000</span> <span class="hljs-number">1000</span> 
 Number of groups: <span class="hljs-number">1</span> 
 Groups names: gastrulation 
 Number of samples (per group): <span class="hljs-number">2177</span> 
 Number of factors: <span class="hljs-number">9</span>
</code></pre>
<p>&#x89C2;&#x5BDF;&#x5355;&#x7EC6;&#x80DE;&#x5728;&#x591A;&#x7EC4;&#x5B66;&#x4E2D;&#x7684;&#x5206;&#x5E03;</p>
<pre><code class="lang-R">plot_data_overview(model)
</code></pre>
<p><img src="https://raw.githubusercontent.com/Starlitnightly/bioinformatic_galaxy/master/img/download.png" alt="download" style="zoom:50%;"></p>
<h3 id="14-&#x6A21;&#x578B;&#x91CD;&#x5B9A;&#x4E49;">1.4 &#x6A21;&#x578B;&#x91CD;&#x5B9A;&#x4E49;</h3>
<p>&#x5728;&#x8FD9;&#x91CC;&#xFF0C;&#x6211;&#x4EEC;&#x9700;&#x8981;&#x5BF9;&#x6A21;&#x578B;&#x7684;&#x4E00;&#x4E9B;&#x53C2;&#x6570;&#x8DDF;&#x53D8;&#x91CF;&#x8FDB;&#x884C;&#x91CD;&#x65B0;&#x5B9A;&#x4E49;</p>
<pre><code class="lang-R">sample_metadata_filt &lt;- sample_metadata %&gt;% 
  setkey(sample) %&gt;% .[samples_names(model)] %&gt;%
  .[,lineage10x_2:=stringr::str_replace_all(lineage10x_2,<span class="hljs-string">&quot;_&quot;</span>,<span class="hljs-string">&quot; &quot;</span>)]

<span class="hljs-comment">#&#x91CD;&#x65B0;&#x5B9A;&#x4E49;&#x7EC4;&#x540D;</span>
opts$views_names &lt;- c(
  <span class="hljs-string">&quot;acc_H3K27ac_distal_E7.5_Mes_intersect12&quot;</span>=<span class="hljs-string">&quot;Mesoderm enhancers (acc)&quot;</span>,
  <span class="hljs-string">&quot;acc_H3K27ac_distal_E7.5_Ect_intersect12&quot;</span>=<span class="hljs-string">&quot;Ectoderm enhancers (acc)&quot;</span>,
  <span class="hljs-string">&quot;acc_H3K27ac_distal_E7.5_End_intersect12&quot;</span>=<span class="hljs-string">&quot;Endoderm enhancers (acc)&quot;</span>,
  <span class="hljs-string">&quot;acc_prom_2000_2000&quot;</span>=<span class="hljs-string">&quot;Acc CGI Promoters&quot;</span>,
  <span class="hljs-string">&quot;met_H3K27ac_distal_E7.5_Mes_intersect12&quot;</span>=<span class="hljs-string">&quot;Mesoderm enhancers (met)&quot;</span>,
  <span class="hljs-string">&quot;met_H3K27ac_distal_E7.5_Ect_intersect12&quot;</span>=<span class="hljs-string">&quot;Ectoderm enhancers (met)&quot;</span>,
  <span class="hljs-string">&quot;met_H3K27ac_distal_E7.5_End_intersect12&quot;</span>=<span class="hljs-string">&quot;Endoderm enhancers (met)&quot;</span>,
  <span class="hljs-string">&quot;met_prom_2000_2000&quot;</span>=<span class="hljs-string">&quot;Met Promoters&quot;</span>,
  <span class="hljs-string">&quot;rna&quot;</span> = <span class="hljs-string">&quot;RNA expression&quot;</span>
)

views_names(model) &lt;- stringr::str_replace_all(views_names(model), opts$views_names)
</code></pre>
<h3 id="15-&#x8BA1;&#x7B97;&#x591A;&#x7EC4;&#x5B66;&#x56E0;&#x5B50;">1.5 &#x8BA1;&#x7B97;&#x591A;&#x7EC4;&#x5B66;&#x56E0;&#x5B50;</h3>
<p>&#x6211;&#x4EEC;&#x8BA1;&#x7B97;MOFA&#x6A21;&#x578B;&#x4E2D;&#x5BF9;&#x4E8E;&#x6BCF;&#x4E00;&#x4E2A;&#x6837;&#x672C;&#x7684;MOFA&#x56E0;&#x5B50;&#xFF0C;&#x5728;&#x4E0D;&#x540C;&#x7684;&#x7EC4;&#x5B66;&#x6570;&#x636E;&#x4E2D;&#x7684;&#x65B9;&#x5DEE;&#x6BD4;&#x4F8B;</p>
<pre><code class="lang-R">r2 &lt;- calculate_variance_explained_per_sample(model)$gastrulation
factors &lt;- r2[,<span class="hljs-string">&quot;RNA&quot;</span>]&gt;<span class="hljs-number">0.005</span>
<span class="hljs-comment">#model &lt;- subset_factors(model, which(factors))</span>
factors_names(model) &lt;- paste(<span class="hljs-string">&quot;Factor&quot;</span>,<span class="hljs-number">1</span>:get_dimensions(model)[[<span class="hljs-string">&quot;K&quot;</span>]], sep=<span class="hljs-string">&quot; &quot;</span>)
</code></pre>
<p>&#x56E0;&#x5B50;&#x53EF;&#x89C6;&#x5316;</p>
<pre><code>plot_variance_explained(model)
</code></pre><p><img src="https://raw.githubusercontent.com/Starlitnightly/bioinformatic_galaxy/master/img/%E4%B8%8B%E8%BD%BD%20(2).png" alt="&#x4E0B;&#x8F7D; (2)" style="zoom:50%;"></p>
<p>&#x53EF;&#x4EE5;&#x770B;&#x5230;&#xFF0C;&#x56E0;&#x5B50;1&#xFF0C;2&#xFF0C;5&#xFF0C;8&#x53EF;&#x80FD;&#x662F;&#x9020;&#x6210;&#x7EC4;&#x5B66;&#x4E2D;&#x6837;&#x672C;&#x6570;&#x636E;&#x53D8;&#x5316;&#x7684;&#x6F5C;&#x5728;&#x56E0;&#x7D20;</p>
<p>&#x6211;&#x4EEC;&#x8FDB;&#x4E00;&#x6B65;&#x89C2;&#x5BDF;&#x8FD9;&#x56DB;&#x4E2A;&#x56E0;&#x5B50;&#x5728;&#x4E0D;&#x540C;&#x7EC6;&#x80DE;&#xFF08;&#x6837;&#x672C;&#xFF09;&#x4E2D;&#x7684;&#x8868;&#x73B0;</p>
<pre><code class="lang-R">p &lt;- plot_factor(model, 
  factors=c(<span class="hljs-number">1</span>,<span class="hljs-number">2</span>,<span class="hljs-number">5</span>,<span class="hljs-number">8</span>), 
  color_by=sample_metadata_filt$lineage10x_2,
  <span class="hljs-comment"># shape_by=sample_metadata_filt$stage,</span>
  dot_size = <span class="hljs-number">1</span>
)
p &lt;- p + scale_color_manual(values=opts$colors)

<span class="hljs-comment"># pdf(paste0(io$outdir,&quot;/pdf/scatterplot_F1vsF2.pdf&quot;), useDingbats = F, onefile = F, width=7, height=4)</span>
print(p)
</code></pre>
<p><img src="https://raw.githubusercontent.com/Starlitnightly/bioinformatic_galaxy/master/img/%E4%B8%8B%E8%BD%BD%20(3" alt="&#x4E0B;&#x8F7D; (3)">.png)</p>
<p>&#x81F3;&#x6B64;&#xFF0C;&#x6211;&#x4EEC;&#x53EF;&#x4EE5;&#x786E;&#x5B9A;&#x56E0;&#x5B50;1&#x4E0E;Mesoderm&#xFF08;&#x4E2D;&#x80DA;&#x5C42;&#xFF09;&#x53CA;Ectoderm&#xFF08;&#x5916;&#x80DA;&#x5C42;&#xFF09;&#x76F8;&#x5173;&#xFF0C;&#x56E0;&#x5B50;2&#x4E0E;Endoderm&#xFF08;&#x5185;&#x80DA;&#x5C42;&#xFF09;&#x76F8;&#x5173;</p>
<h2 id="2&#x3001;&#x7EC6;&#x80DE;&#x547D;&#x8FD0;&#x53EF;&#x89C6;&#x5316;">2&#x3001;&#x7EC6;&#x80DE;&#x547D;&#x8FD0;&#x53EF;&#x89C6;&#x5316;</h2>
<h3 id="21-&#x52A0;&#x8F7D;&#x4F9D;&#x8D56;">2.1 &#x52A0;&#x8F7D;&#x4F9D;&#x8D56;</h3>
<pre><code class="lang-R"><span class="hljs-keyword">library</span>(umap)
<span class="hljs-keyword">library</span>(Rtsne)
<span class="hljs-keyword">library</span>(irlba)
</code></pre>
<h3 id="22-&#x5B9A;&#x4E49;&#x51FD;&#x6570;">2.2 &#x5B9A;&#x4E49;&#x51FD;&#x6570;</h3>
<pre><code class="lang-R">theme_pub &lt;- <span class="hljs-keyword">function</span>() {
  theme_classic() +
  theme(
    legend.position = <span class="hljs-string">&quot;right&quot;</span>,
    legend.title = element_blank(),
    legend.text = element_text(size=rel(<span class="hljs-number">1.2</span>)),
    axis.title = element_text(size=rel(<span class="hljs-number">1.2</span>), color=<span class="hljs-string">&quot;black&quot;</span>),
    axis.text = element_blank(),
    axis.ticks = element_blank()
  )
}
</code></pre>
<h3 id="23-umap&#x53C2;&#x6570;&#x8BBE;&#x7F6E;">2.3 umap&#x53C2;&#x6570;&#x8BBE;&#x7F6E;</h3>
<pre><code class="lang-R">algorithms &lt;- c(<span class="hljs-string">&quot;umap&quot;</span>)

<span class="hljs-comment"># umap.defaults$n_neighbors &lt;- 20</span>
<span class="hljs-comment"># umap.defaults$min_dist &lt;- 0.6</span>

umap.defaults$n_neighbors &lt;- <span class="hljs-number">25</span>
umap.defaults$min_dist &lt;- <span class="hljs-number">0.55</span>
</code></pre>
<h3 id="24-&#x7ED8;&#x5236;&#x4E0D;&#x540C;&#x79CD;&#x7C7B;&#x7684;&#x7EC6;&#x80DE;&#x5206;&#x5E03;&#x56FE;">2.4 &#x7ED8;&#x5236;&#x4E0D;&#x540C;&#x79CD;&#x7C7B;&#x7684;&#x7EC6;&#x80DE;&#x5206;&#x5E03;&#x56FE;</h3>
<pre><code class="lang-R"><span class="hljs-comment"># Fetch factors</span>
factors &lt;- c(<span class="hljs-number">1</span>,<span class="hljs-number">2</span>,<span class="hljs-number">4</span>,<span class="hljs-number">8</span>)
Z &lt;- get_factors(model,factors=c(<span class="hljs-number">1</span>,<span class="hljs-number">2</span>,<span class="hljs-number">4</span>,<span class="hljs-number">8</span>))$gastrulation 

<span class="hljs-keyword">for</span> (algorithm <span class="hljs-keyword">in</span> algorithms) {

  set.seed(<span class="hljs-number">42</span>)
  <span class="hljs-keyword">if</span> (algorithm==<span class="hljs-string">&quot;tsne&quot;</span>) {
    tsne &lt;- Rtsne(Z, check_duplicates=<span class="hljs-literal">FALSE</span>, pca=<span class="hljs-literal">FALSE</span>, theta=<span class="hljs-number">0.5</span>, dims=<span class="hljs-number">2</span>)
    Z.out &lt;- tsne$Y
  } <span class="hljs-keyword">else</span> <span class="hljs-keyword">if</span> (algorithm==<span class="hljs-string">&quot;umap&quot;</span>) {
    umap.out &lt;- umap(Z, config = umap.defaults)
    Z.out &lt;- umap.out$layout
  }

  <span class="hljs-comment"># Flip a factor  </span>
  Z.out[,<span class="hljs-number">2</span>] &lt;- -Z.out[,<span class="hljs-number">2</span>]

  to.plot &lt;- Z.out %&gt;% as.data.table %&gt;% .[,sample:=rownames(Z)] %&gt;%
      merge(sample_metadata_filt, by=<span class="hljs-string">&quot;sample&quot;</span>)

  p1 &lt;- ggplot(to.plot, aes(x=V1, y=V2, color=lineage10x_2)) +
    geom_point(alpha=<span class="hljs-number">0.7</span>, size=<span class="hljs-number">2.0</span>) +
    scale_color_manual(values=opts$colors) +
    guides(colour = guide_legend(override.aes = list(size=<span class="hljs-number">3</span>))) +
    labs(x=<span class="hljs-string">&quot;UMAP Dimension 1&quot;</span>, y=<span class="hljs-string">&quot;UMAP Dimension 2&quot;</span>) +
    theme_pub()

  <span class="hljs-comment"># pdf(sprintf(&quot;%s/MOFA_%s.pdf&quot;,io$outdir,algorithm), width=5, height=7, useDingbats = F)</span>
  print(p1)
  <span class="hljs-comment"># dev.off()</span>

  <span class="hljs-comment"># Save coordinates</span>
  <span class="hljs-comment"># fwrite(to.plot[,c(&quot;sample&quot;,&quot;V1&quot;,&quot;V2&quot;)], sprintf(&quot;%s/%s_coordinates.txt&quot;,io$outdir,algorithm))</span>
}
</code></pre>
<p><img src="https://raw.githubusercontent.com/Starlitnightly/bioinformatic_galaxy/master/img/%E4%B8%8B%E8%BD%BD%20(5).png" alt="&#x4E0B;&#x8F7D; (5)" style="zoom:50%;"></p>
<h3 id="25-&#x63A8;&#x65AD;&#x7F3A;&#x5931;&#x503C;">2.5 &#x63A8;&#x65AD;&#x7F3A;&#x5931;&#x503C;</h3>
<p>&#x6211;&#x4EEC;&#x4F7F;&#x7528;impute&#x63A8;&#x65AD;&#x5728;&#x5355;&#x7EC6;&#x80DE;&#x4E2D;&#x76EE;&#x6807;&#x503C;&#x7684;&#x7F3A;&#x5931;&#x60C5;&#x51B5;</p>
<pre><code class="lang-R">model &lt;- impute(model)
</code></pre>
<h3 id="26-&#x7ED8;&#x5236;&#x7532;&#x57FA;&#x5316;&#x56FE;&#x8C31;">2.6 &#x7ED8;&#x5236;&#x7532;&#x57FA;&#x5316;&#x56FE;&#x8C31;</h3>
<h4 id="261-&#x5916;&#x80DA;&#x5C42;&#x589E;&#x5F3A;&#x5B50;">2.6.1 &#x5916;&#x80DA;&#x5C42;&#x589E;&#x5F3A;&#x5B50;</h4>
<pre><code class="lang-R"><span class="hljs-comment">#</span>
view &lt;- <span class="hljs-string">&quot;Ectoderm enhancers (met)&quot;</span>
factor &lt;- <span class="hljs-number">1</span>

<span class="hljs-comment"># Select top 50 enhancers with largest loading (Factor 1 for mesoderm enhancers, Factor 2 for endoderm enhancers)</span>
tmp &lt;- names(tail(sort(abs(get_weights(model, views=view, factor=factor)[[<span class="hljs-number">1</span>]][,<span class="hljs-number">1</span>])), n=<span class="hljs-number">50</span>))
<span class="hljs-keyword">if</span> (length(model@imputed_data)&gt;<span class="hljs-number">0</span>) {
  met &lt;- colMeans(model@imputed_data[[view]]$gastrulation[tmp,], na.rm=<span class="hljs-literal">T</span>)
} <span class="hljs-keyword">else</span> {
  met &lt;- colMeans(model@data[[view]]$gastrulation[tmp,], na.rm=<span class="hljs-literal">T</span>)
}

<span class="hljs-comment"># for better visualisation</span>
met[met&lt;((-<span class="hljs-number">6</span>) )] &lt;- (-<span class="hljs-number">6</span>) 
met[met&gt;(<span class="hljs-number">5</span>)] &lt;- <span class="hljs-number">5</span>

<span class="hljs-comment"># Convert M-values to B-values</span>
met &lt;- <span class="hljs-number">100</span>*<span class="hljs-number">2</span>**met/(<span class="hljs-number">1</span>+<span class="hljs-number">2</span>**met)

foo &lt;- to.plot %&gt;% merge(
  data.table(sample = samples_names(model)$gastrulation, met = met), by=<span class="hljs-string">&quot;sample&quot;</span>
)

p &lt;- ggplot(foo, aes(x=V1, y=V2, color=met)) +
  geom_point(alpha=<span class="hljs-number">0.7</span>, size=<span class="hljs-number">2.0</span>) +
  scale_colour_gradientn(colours = brewer.pal(<span class="hljs-number">9</span>, <span class="hljs-string">&quot;OrRd&quot;</span>)) +
  labs(x=<span class="hljs-string">&quot;UMAP Dimension 1&quot;</span>, y=<span class="hljs-string">&quot; UMAP Dimension 2&quot;</span>) +
  theme_pub() 
<span class="hljs-comment"># print(p)</span>

<span class="hljs-comment"># pdf(sprintf(&quot;%s/MOFA_%s_%d.pdf&quot;,io$outdir,view,factor), width=6, height=6, useDingbats = F)</span>
print(p)
</code></pre>
<p><img src="https://raw.githubusercontent.com/Starlitnightly/bioinformatic_galaxy/master/img/%E4%B8%8B%E8%BD%BD%20(6).png" alt="&#x4E0B;&#x8F7D; (6)" style="zoom:50%;"></p>
<h4 id="262-&#x4E2D;&#x80DA;&#x5C42;&#x589E;&#x5F3A;&#x5B50;">2.6.2 &#x4E2D;&#x80DA;&#x5C42;&#x589E;&#x5F3A;&#x5B50;</h4>
<pre><code class="lang-R">view &lt;- <span class="hljs-string">&quot;Mesoderm enhancers (met)&quot;</span>
factor &lt;- <span class="hljs-number">1</span>

<span class="hljs-comment"># Select top 50 enhancers with largest loading (Factor 1 for mesoderm enhancers, Factor 2 for endoderm enhancers)</span>
tmp &lt;- names(tail(sort(abs(get_weights(model, views=view, factor=factor)[[<span class="hljs-number">1</span>]][,<span class="hljs-number">1</span>])), n=<span class="hljs-number">50</span>))
<span class="hljs-keyword">if</span> (length(model@imputed_data)&gt;<span class="hljs-number">0</span>) {
  met &lt;- colMeans(model@imputed_data[[view]]$gastrulation[tmp,], na.rm=<span class="hljs-literal">T</span>)
} <span class="hljs-keyword">else</span> {
  met &lt;- colMeans(model@data[[view]]$gastrulation[tmp,], na.rm=<span class="hljs-literal">T</span>)
}

<span class="hljs-comment"># for better visualisation</span>
met[met&lt;((-<span class="hljs-number">6</span>) )] &lt;- (-<span class="hljs-number">6</span>) 
met[met&gt;(<span class="hljs-number">5</span>)] &lt;- <span class="hljs-number">5</span>

<span class="hljs-comment"># Convert M-values to B-values</span>
met &lt;- <span class="hljs-number">100</span>*<span class="hljs-number">2</span>**met/(<span class="hljs-number">1</span>+<span class="hljs-number">2</span>**met)

foo &lt;- to.plot %&gt;% merge(
  data.table(sample = samples_names(model)$gastrulation, met = met), by=<span class="hljs-string">&quot;sample&quot;</span>
)

p &lt;- ggplot(foo, aes(x=V1, y=V2, color=met)) +
  geom_point(alpha=<span class="hljs-number">0.7</span>, size=<span class="hljs-number">2.0</span>) +
  scale_colour_gradientn(colours = brewer.pal(<span class="hljs-number">9</span>, <span class="hljs-string">&quot;OrRd&quot;</span>)) +
  labs(x=<span class="hljs-string">&quot;UMAP Dimension 1&quot;</span>, y=<span class="hljs-string">&quot; UMAP Dimension 2&quot;</span>) +
  theme_pub() 
<span class="hljs-comment"># print(p)</span>

<span class="hljs-comment"># pdf(sprintf(&quot;%s/MOFA_%s_%d.pdf&quot;,io$outdir,view,factor), width=6, height=6, useDingbats = F)</span>
print(p)
<span class="hljs-comment"># dev.off()</span>
</code></pre>
<p><img src="https://raw.githubusercontent.com/Starlitnightly/bioinformatic_galaxy/master/img/%E4%B8%8B%E8%BD%BD%20(7).png" alt="&#x4E0B;&#x8F7D; (7)" style="zoom:50%;"></p>
<h4 id="263-&#x5185;&#x80DA;&#x5C42;&#x589E;&#x5F3A;&#x5B50;">2.6.3 &#x5185;&#x80DA;&#x5C42;&#x589E;&#x5F3A;&#x5B50;</h4>
<pre><code class="lang-R"><span class="hljs-comment">#</span>
view &lt;- <span class="hljs-string">&quot;Endoderm enhancers (met)&quot;</span>
factor &lt;- <span class="hljs-number">2</span>

<span class="hljs-comment"># Select top 50 enhancers with largest loading (Factor 1 for mesoderm enhancers, Factor 2 for endoderm enhancers)</span>
tmp &lt;- names(tail(sort(abs(get_weights(model, views=view, factor=factor)[[<span class="hljs-number">1</span>]][,<span class="hljs-number">1</span>])), n=<span class="hljs-number">50</span>))
<span class="hljs-keyword">if</span> (length(model@imputed_data)&gt;<span class="hljs-number">0</span>) {
  met &lt;- colMeans(model@imputed_data[[view]]$gastrulation[tmp,], na.rm=<span class="hljs-literal">T</span>)
} <span class="hljs-keyword">else</span> {
  met &lt;- colMeans(model@data[[view]]$gastrulation[tmp,], na.rm=<span class="hljs-literal">T</span>)
}

<span class="hljs-comment"># for better visualisation</span>
met[met&lt;((-<span class="hljs-number">6</span>) )] &lt;- (-<span class="hljs-number">6</span>) 
met[met&gt;(<span class="hljs-number">5</span>)] &lt;- <span class="hljs-number">5</span>

<span class="hljs-comment"># Convert M-values to B-values</span>
met &lt;- <span class="hljs-number">100</span>*<span class="hljs-number">2</span>**met/(<span class="hljs-number">1</span>+<span class="hljs-number">2</span>**met)

foo &lt;- to.plot %&gt;% merge(
  data.table(sample = samples_names(model)$gastrulation, met = met), by=<span class="hljs-string">&quot;sample&quot;</span>
)

p &lt;- ggplot(foo, aes(x=V1, y=V2, color=met)) +
  geom_point(alpha=<span class="hljs-number">0.7</span>, size=<span class="hljs-number">2.0</span>) +
  scale_colour_gradientn(colours = brewer.pal(<span class="hljs-number">9</span>, <span class="hljs-string">&quot;OrRd&quot;</span>)) +
  labs(x=<span class="hljs-string">&quot;UMAP Dimension 1&quot;</span>, y=<span class="hljs-string">&quot; UMAP Dimension 2&quot;</span>) +
  theme_pub() 
<span class="hljs-comment"># print(p)</span>

<span class="hljs-comment"># pdf(sprintf(&quot;%s/MOFA_%s_%d.pdf&quot;,io$outdir,view,factor), width=6, height=6, useDingbats = F)</span>
print(p)
</code></pre>
<p><img src="https://raw.githubusercontent.com/Starlitnightly/bioinformatic_galaxy/master/img/%E4%B8%8B%E8%BD%BD%20(8).png" alt="&#x4E0B;&#x8F7D; (8)" style="zoom:50%;"></p>
<h3 id="27-&#x7ED8;&#x5236;&#x67D3;&#x8272;&#x8D28;&#x5F00;&#x53D1;&#x6027;&#x56FE;&#x8C31;">2.7 &#x7ED8;&#x5236;&#x67D3;&#x8272;&#x8D28;&#x5F00;&#x53D1;&#x6027;&#x56FE;&#x8C31;</h3>
<h4 id="271-&#x5916;&#x80DA;&#x5C42;&#x589E;&#x5F3A;&#x5B50;">2.7.1 &#x5916;&#x80DA;&#x5C42;&#x589E;&#x5F3A;&#x5B50;</h4>
<pre><code class="lang-R"><span class="hljs-comment">#  &quot;acc_H3K27ac_distal_E7.5_Mes_intersect12&quot;=&quot;Mesoderm enhancers (acc)&quot;,</span>
<span class="hljs-comment">#&quot;acc_H3K27ac_distal_E7.5_Ect_intersect12&quot;=&quot;Ectoderm enhancers (acc)&quot;,</span>
<span class="hljs-comment">#  &quot;acc_H3K27ac_distal_E7.5_End_intersect12&quot;=&quot;Endoderm enhancers (acc)&quot;,</span>
<span class="hljs-comment">#</span>

view &lt;- <span class="hljs-string">&quot;Ectoderm enhancers (acc)&quot;</span>
factor &lt;- <span class="hljs-number">1</span>

<span class="hljs-comment"># Select top 50 enhancers with largest loading (Factor 1 for mesoderm enhancers, Factor 2 for endoderm enhancers)</span>
tmp &lt;- names(tail(sort(abs(get_weights(model, views=view, factor=factor)[[<span class="hljs-number">1</span>]][,<span class="hljs-number">1</span>])), n=<span class="hljs-number">50</span>))
<span class="hljs-keyword">if</span> (length(model@imputed_data)&gt;<span class="hljs-number">0</span>) {
  acc &lt;- colMeans(model@imputed_data[[view]]$gastrulation[tmp,], na.rm=<span class="hljs-literal">T</span>)
} <span class="hljs-keyword">else</span> {
  acc &lt;- colMeans(model@data[[view]]$gastrulation[tmp,], na.rm=<span class="hljs-literal">T</span>)
}



<span class="hljs-comment"># For better visualisation</span>
acc[acc&lt;(-<span class="hljs-number">1.5</span>)] &lt;- (-<span class="hljs-number">1.5</span>)
acc[acc&gt;(<span class="hljs-number">0.5</span>)] &lt;- <span class="hljs-number">0.5</span>

<span class="hljs-comment"># Convert M-values to B-values</span>
acc &lt;- <span class="hljs-number">100</span>*<span class="hljs-number">2</span>**acc/(<span class="hljs-number">1</span>+<span class="hljs-number">2</span>**acc)

foo &lt;- to.plot %&gt;% merge(
  data.table(sample = samples_names(model)$gastrulation, acc = acc), by=<span class="hljs-string">&quot;sample&quot;</span>
)

p &lt;- ggplot(foo, aes(x=V1, y=V2, color=acc)) +
  geom_point(alpha=<span class="hljs-number">0.7</span>, size=<span class="hljs-number">1.5</span>) +
  scale_colour_gradientn(colours = rev(brewer.pal(<span class="hljs-number">9</span>, <span class="hljs-string">&quot;Blues&quot;</span>))) +
  labs(x=<span class="hljs-string">&quot;UMAP Dimension 1&quot;</span>, y=<span class="hljs-string">&quot; UMAP Dimension 2&quot;</span>) +
  theme_pub() 

<span class="hljs-comment"># pdf(sprintf(&quot;%s/MOFA_%s_%d.pdf&quot;,io$outdir,view,factor), width=6, height=6, useDingbats = F)</span>
print(p)
</code></pre>
<p><img src="https://raw.githubusercontent.com/Starlitnightly/bioinformatic_galaxy/master/img/%E4%B8%8B%E8%BD%BD%20(9).png" alt="&#x4E0B;&#x8F7D; (9)" style="zoom:50%;"></p>
<h4 id="272-&#x4E2D;&#x80DA;&#x5C42;&#x589E;&#x5F3A;&#x5B50;">2.7.2 &#x4E2D;&#x80DA;&#x5C42;&#x589E;&#x5F3A;&#x5B50;</h4>
<pre><code class="lang-R"><span class="hljs-comment">#  &quot;acc_H3K27ac_distal_E7.5_Mes_intersect12&quot;=&quot;Mesoderm enhancers (acc)&quot;,</span>
<span class="hljs-comment">#&quot;acc_H3K27ac_distal_E7.5_Ect_intersect12&quot;=&quot;Ectoderm enhancers (acc)&quot;,</span>
<span class="hljs-comment">#  &quot;acc_H3K27ac_distal_E7.5_End_intersect12&quot;=&quot;Endoderm enhancers (acc)&quot;,</span>
<span class="hljs-comment">#</span>

view &lt;- <span class="hljs-string">&quot;Mesoderm enhancers (acc)&quot;</span>
factor &lt;- <span class="hljs-number">1</span>

<span class="hljs-comment"># Select top 50 enhancers with largest loading (Factor 1 for mesoderm enhancers, Factor 2 for endoderm enhancers)</span>
tmp &lt;- names(tail(sort(abs(get_weights(model, views=view, factor=factor)[[<span class="hljs-number">1</span>]][,<span class="hljs-number">1</span>])), n=<span class="hljs-number">50</span>))
<span class="hljs-keyword">if</span> (length(model@imputed_data)&gt;<span class="hljs-number">0</span>) {
  acc &lt;- colMeans(model@imputed_data[[view]]$gastrulation[tmp,], na.rm=<span class="hljs-literal">T</span>)
} <span class="hljs-keyword">else</span> {
  acc &lt;- colMeans(model@data[[view]]$gastrulation[tmp,], na.rm=<span class="hljs-literal">T</span>)
}



<span class="hljs-comment"># For better visualisation</span>
acc[acc&lt;(-<span class="hljs-number">1.5</span>)] &lt;- (-<span class="hljs-number">1.5</span>)
acc[acc&gt;(<span class="hljs-number">0.5</span>)] &lt;- <span class="hljs-number">0.5</span>

<span class="hljs-comment"># Convert M-values to B-values</span>
acc &lt;- <span class="hljs-number">100</span>*<span class="hljs-number">2</span>**acc/(<span class="hljs-number">1</span>+<span class="hljs-number">2</span>**acc)

foo &lt;- to.plot %&gt;% merge(
  data.table(sample = samples_names(model)$gastrulation, acc = acc), by=<span class="hljs-string">&quot;sample&quot;</span>
)

p &lt;- ggplot(foo, aes(x=V1, y=V2, color=acc)) +
  geom_point(alpha=<span class="hljs-number">0.7</span>, size=<span class="hljs-number">1.5</span>) +
  scale_colour_gradientn(colours = rev(brewer.pal(<span class="hljs-number">9</span>, <span class="hljs-string">&quot;Blues&quot;</span>))) +
  labs(x=<span class="hljs-string">&quot;UMAP Dimension 1&quot;</span>, y=<span class="hljs-string">&quot; UMAP Dimension 2&quot;</span>) +
  theme_pub() 

<span class="hljs-comment"># pdf(sprintf(&quot;%s/MOFA_%s_%d.pdf&quot;,io$outdir,view,factor), width=6, height=6, useDingbats = F)</span>
print(p)
</code></pre>
<p><img src="https://raw.githubusercontent.com/Starlitnightly/bioinformatic_galaxy/master/img/%E4%B8%8B%E8%BD%BD%20(10).png" alt="&#x4E0B;&#x8F7D; (10)" style="zoom:50%;"></p>
<h4 id="273-&#x5185;&#x80DA;&#x5C42;&#x589E;&#x5F3A;&#x5B50;">2.7.3 &#x5185;&#x80DA;&#x5C42;&#x589E;&#x5F3A;&#x5B50;</h4>
<pre><code class="lang-R"><span class="hljs-comment">#  &quot;acc_H3K27ac_distal_E7.5_Mes_intersect12&quot;=&quot;Mesoderm enhancers (acc)&quot;,</span>
<span class="hljs-comment">#&quot;acc_H3K27ac_distal_E7.5_Ect_intersect12&quot;=&quot;Ectoderm enhancers (acc)&quot;,</span>
<span class="hljs-comment">#  &quot;acc_H3K27ac_distal_E7.5_End_intersect12&quot;=&quot;Endoderm enhancers (acc)&quot;,</span>
<span class="hljs-comment">#</span>

view &lt;- <span class="hljs-string">&quot;Endoderm enhancers (acc)&quot;</span>
factor &lt;- <span class="hljs-number">2</span>

<span class="hljs-comment"># Select top 50 enhancers with largest loading (Factor 1 for mesoderm enhancers, Factor 2 for endoderm enhancers)</span>
tmp &lt;- names(tail(sort(abs(get_weights(model, views=view, factor=factor)[[<span class="hljs-number">1</span>]][,<span class="hljs-number">1</span>])), n=<span class="hljs-number">50</span>))
<span class="hljs-keyword">if</span> (length(model@imputed_data)&gt;<span class="hljs-number">0</span>) {
  acc &lt;- colMeans(model@imputed_data[[view]]$gastrulation[tmp,], na.rm=<span class="hljs-literal">T</span>)
} <span class="hljs-keyword">else</span> {
  acc &lt;- colMeans(model@data[[view]]$gastrulation[tmp,], na.rm=<span class="hljs-literal">T</span>)
}



<span class="hljs-comment"># For better visualisation</span>
acc[acc&lt;(-<span class="hljs-number">1.5</span>)] &lt;- (-<span class="hljs-number">1.5</span>)
acc[acc&gt;(<span class="hljs-number">0.5</span>)] &lt;- <span class="hljs-number">0.5</span>

<span class="hljs-comment"># Convert M-values to B-values</span>
acc &lt;- <span class="hljs-number">100</span>*<span class="hljs-number">2</span>**acc/(<span class="hljs-number">1</span>+<span class="hljs-number">2</span>**acc)

foo &lt;- to.plot %&gt;% merge(
  data.table(sample = samples_names(model)$gastrulation, acc = acc), by=<span class="hljs-string">&quot;sample&quot;</span>
)

p &lt;- ggplot(foo, aes(x=V1, y=V2, color=acc)) +
  geom_point(alpha=<span class="hljs-number">0.7</span>, size=<span class="hljs-number">1.5</span>) +
  scale_colour_gradientn(colours = rev(brewer.pal(<span class="hljs-number">9</span>, <span class="hljs-string">&quot;Blues&quot;</span>))) +
  labs(x=<span class="hljs-string">&quot;UMAP Dimension 1&quot;</span>, y=<span class="hljs-string">&quot; UMAP Dimension 2&quot;</span>) +
  theme_pub() 

<span class="hljs-comment"># pdf(sprintf(&quot;%s/MOFA_%s_%d.pdf&quot;,io$outdir,view,factor), width=6, height=6, useDingbats = F)</span>
print(p)
</code></pre>
<p><img src="https://raw.githubusercontent.com/Starlitnightly/bioinformatic_galaxy/master/img/%E4%B8%8B%E8%BD%BD%20(11).png" alt="&#x4E0B;&#x8F7D; (11)" style="zoom:50%;"></p>
<h2 id="3&#x3001;&#x540E;&#x8BDD;">3&#x3001;&#x540E;&#x8BDD;</h2>
<p>&#x5230;&#x8FD9;&#x91CC;&#xFF0C;Figure3b&#x5C31;&#x5B8C;&#x5168;&#x590D;&#x73B0;&#x5B8C;&#x4E86;&#xFF0C;&#x4E2D;&#x95F4;&#x8FD8;&#x662F;&#x6709;&#x5F88;&#x591A;&#x5F88;&#x6709;&#x610F;&#x601D;&#x7684;&#x70B9;&#x3002;&#x7531;&#x4E8E;&#x6211;&#x8FD9;&#x91CC;&#x53EA;&#x6709;&#x4E00;&#x7EC4;&#xFF0C;&#x6240;&#x4EE5;MOFA&#x8BA1;&#x7B97;&#x51FA;&#x6765;&#x7684;&#x56E0;&#x5B50;&#x5176;&#x5B9E;&#x662F;&#x4E0D;&#x540C;&#x7EC4;&#x5B66;&#x6570;&#x636E;&#x53EF;&#x53D8;&#x6027;&#x7684;&#x6765;&#x6E90;&#xFF0C;&#x800C;&#x5E76;&#x4E0D;&#x662F;&#x7EC4;&#x95F4;&#x5DEE;&#x5F02;&#xFF0C;&#x8FDB;&#x800C;&#x501F;&#x52A9;&#x7EC4;&#x5B66;&#x6570;&#x636E;&#x7684;&#x5DEE;&#x5F02;&#x6765;&#x786E;&#x5B9A;</p>
<p>&#x6CE8;&#xFF1A;&#x672C;&#x7814;&#x7A76;&#x8FD0;&#x7528;MOFA&#x5206;&#x6790;&#x7684;&#x539F;&#x56E0;&#x662F;&#xFF0C;&#x6BCF;&#x4E2A;&#x7EC4;&#x5B66;&#x5C42;&#x7684;&#x6570;&#x636E;&#x90FD;&#x6709;&#x4E00;&#x5927;&#x5806;&#x7EC6;&#x80DE;&#xFF0C;&#x4F5C;&#x8005;&#x7684;&#x76EE;&#x7684;&#x662F;&#x627E;&#x51FA;&#x5BFC;&#x81F4;&#x8FD9;&#x4E00;&#x5927;&#x5806;&#x7EC6;&#x80DE;&#x4E0D;&#x540C;&#x7684;&#x539F;&#x56E0;&#xFF0C;&#x4E5F;&#x5C31;&#x662F;&#x6240;&#x8C13;&#x7684;&#x65B9;&#x5DEE;&#x6700;&#x5927;&#x3002;MOFA&#x5E94;&#x7528;&#x5728;&#x627E;&#x6709;&#x75C5;&#x6CA1;&#x75C5;&#x7684;&#x4EBA;&#x7684;hub&#x57FA;&#x56E0;&#x53EF;&#x80FD;&#x4F1A;&#x6709;&#x4E9B;&#x9EBB;&#x70E6;&#xFF0C;&#x56E0;&#x4E3A;&#x53EF;&#x80FD;hub&#x57FA;&#x56E0;&#x5728;&#x6709;&#x75C5;&#x7684;&#x4EBA;&#x4F53;&#x5185;&#x8868;&#x8FBE;&#x90FD;&#x5F88;&#x9AD8;&#xFF0C;&#x5728;&#x6CA1;&#x75C5;&#x7684;&#x4EBA;&#x4F53;&#x5185;&#x8868;&#x8FBE;&#x90FD;&#x5F88;&#x4F4E;&#xFF0C;&#x8FD9;&#x79CD;&#x60C5;&#x51B5;&#x4E0B;&#xFF0C;&#x7EC4;&#x5185;&#x7EC4;&#x5B66;&#x6570;&#x636E;&#x7684;&#x65B9;&#x5DEE;&#x5C31;&#x4F1A;&#x5F88;&#x5C0F;&#x3002;&#x4E0D;&#x8FC7;&#x4E5F;&#x4E0D;&#x662F;&#x4E0D;&#x53EF;&#x4EE5;&#xFF0C;&#x628A;&#x6709;&#x75C5;&#x6CA1;&#x75C5;&#x7684;&#x6570;&#x636E;&#x6DF7;&#x6210;&#x4E00;&#x7EC4;&#x5C31;&#x597D;&#x4E86;&#x3002;</p>

                                
                                </section>
                            
    </div>
    <div class="search-results">
        <div class="has-results">
            
            <h1 class="search-results-title"><span class='search-results-count'></span> results matching "<span class='search-query'></span>"</h1>
            <ul class="search-results-list"></ul>
            
        </div>
        <div class="no-results">
            
            <h1 class="search-results-title">No results matching "<span class='search-query'></span>"</h1>
            
        </div>
    </div>
</div>

                        </div>
                    </div>
                
            </div>

            
                
                <a href="gastrulation_2.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page: 1-2. MOFA模型-Figure3b">
                    <i class="fa fa-angle-left"></i>
                </a>
                
                
            
        
    </div>

    <script>
        var gitbook = gitbook || [];
        gitbook.push(function() {
            gitbook.page.hasChanged({"page":{"title":"1-3. 下游分析-Figure3b","level":"1.2.3","depth":2,"previous":{"title":"1-2. MOFA模型-Figure3b","level":"1.2.2","depth":2,"path":"GASTRULATION/gastrulation_2.md","ref":"GASTRULATION/gastrulation_2.md","articles":[]},"dir":"ltr"},"config":{"gitbook":"*","theme":"default","variables":{},"plugins":["expandable-chapters"],"pluginsConfig":{"expandable-chapters":{},"highlight":{},"search":{},"lunr":{"maxIndexSize":1000000,"ignoreSpecialCharacters":false},"sharing":{"facebook":true,"twitter":true,"google":false,"weibo":false,"instapaper":false,"vk":false,"all":["facebook","google","twitter","weibo","instapaper"]},"fontsettings":{"theme":"white","family":"sans","size":2},"theme-default":{"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"showLevel":false}},"structure":{"langs":"LANGS.md","readme":"README.md","glossary":"GLOSSARY.md","summary":"SUMMARY.md"},"pdf":{"pageNumbers":true,"fontSize":12,"fontFamily":"Arial","paperSize":"a4","chapterMark":"pagebreak","pageBreaksBefore":"/","margin":{"right":62,"left":62,"top":56,"bottom":56}},"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"}},"file":{"path":"GASTRULATION/gastrulation_3.md","mtime":"2021-10-06T20:12:28.652Z","type":"markdown"},"gitbook":{"version":"3.2.3","time":"2021-10-06T21:05:52.176Z"},"basePath":"..","book":{"language":""}});
        });
    </script>
</div>

        
    <script src="../gitbook/gitbook.js"></script>
    <script src="../gitbook/theme.js"></script>
    
        
        <script src="../gitbook/gitbook-plugin-expandable-chapters/expandable-chapters.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search-engine.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/lunr.min.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/search-lunr.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-sharing/buttons.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-fontsettings/fontsettings.js"></script>
        
    

    </body>
</html>

